// @antv/xflow-core/lib/command-contributions/components/context.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useIsNodeSelected = exports.getEdgeReactComponent = exports.getNodeReactComponent = exports.useAppContext = exports.AppContext = void 0;
var react_1 = __importDefault(require("react"));
exports.AppContext = react_1.default.createContext({
    cell: null,
    x6Graph: null,
    commands: null,
    modelService: null,
});
var useAppContext = function () { return react_1.default.useContext(exports.AppContext); };
exports.useAppContext = useAppContext;
function getNodeReactComponent(Component, commands, modelService) {
    return function (x6Node) {
        /* eslint-disable-next-line  */
        var x6Graph = this;
        var nodeData = x6Node === null || x6Node === void 0 ? void 0 : x6Node.getData();
        var size = x6Node === null || x6Node === void 0 ? void 0 : x6Node.getSize();
        var position = x6Node === null || x6Node === void 0 ? void 0 : x6Node.getPosition();
        var WrappedComponent = function () {
            return (react_1.default.createElement(exports.AppContext.Provider, { value: {
                    cell: x6Node,
                    x6Graph: x6Graph,
                    commands: commands,
                    modelService: modelService,
                } }, react_1.default.createElement(Component, {
                cell: x6Node,
                data: nodeData,
                size: size,
                position: position,
            })));
        };
        return react_1.default.createElement(WrappedComponent, null);
    };
}
exports.getNodeReactComponent = getNodeReactComponent;
function getEdgeReactComponent(Component, commands, modelService) {
    return function (x6Edge) {
        /* eslint-disable-next-line  */
        var x6Graph = this;
        var edgeData = x6Edge === null || x6Edge === void 0 ? void 0 : x6Edge.getData();
        var WrappedComponent = function () {
            return (react_1.default.createElement(exports.AppContext.Provider, { value: {
                    cell: x6Edge,
                    x6Graph: x6Graph,
                    commands: commands,
                    modelService: modelService,
                } }, react_1.default.createElement(Component, {
                data: edgeData,
                cell: x6Edge,
            })));
        };
        return react_1.default.createElement(WrappedComponent, null);
    };
}
exports.getEdgeReactComponent = getEdgeReactComponent;
/** 获取节点选中状态 */
var useIsNodeSelected = function () {
    var _a = react_1.default.useState(false), isSelected = _a[0], setIsSelected = _a[1];
    var _b = (0, exports.useAppContext)(), x6Graph = _b.x6Graph, cell = _b.cell;
    react_1.default.useEffect(function () {
        var handler = function (_a) {
            var added = _a.added, removed = _a.removed;
            var isAdded = added
                .filter(function (selectedCell) { return selectedCell.isNode(); })
                .some(function (node) {
                node.id === cell.id;
            });
            if (isAdded) {
                cell.setData(__assign(__assign({}, cell.getData()), { isNodeSelected: true }));
                setIsSelected(true);
                return;
            }
            var isRemoved = removed
                .filter(function (selectedCell) { return selectedCell.isNode(); })
                .some(function (node) {
                node.id === cell.id;
            });
            if (isRemoved) {
                cell.setData(__assign(__assign({}, cell.getData()), { isNodeSelected: false }));
                setIsSelected(false);
                return;
            }
        };
        x6Graph.on('selection:changed', handler);
        return function () {
            x6Graph.off('selection:changed', handler);
        };
        /* eslint-disable-next-line  */
    }, [x6Graph]);
    return [isSelected];
};
exports.useIsNodeSelected = useIsNodeSelected;
