// @antv/xflow-core/lib/model-service/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createModelServiceConfig = exports.ModelRegistry = exports.ModelServiceRegistry = exports.ModelServiceConfig = exports.IModelContribution = exports.IModelService = exports.MODELS = void 0;
var interface_1 = require("./interface");
Object.defineProperty(exports, "IModelService", { enumerable: true, get: function () { return interface_1.IModelService; } });
Object.defineProperty(exports, "IModelContribution", { enumerable: true, get: function () { return interface_1.IModelContribution; } });
var config_1 = require("./config");
Object.defineProperty(exports, "ModelServiceConfig", { enumerable: true, get: function () { return config_1.ModelServiceConfig; } });
var model_registry_1 = require("./model-registry");
Object.defineProperty(exports, "ModelRegistry", { enumerable: true, get: function () { return model_registry_1.ModelRegistry; } });
var components_1 = require("./components");
Object.defineProperty(exports, "ModelServiceRegistry", { enumerable: true, get: function () { return components_1.ModelServiceRegistry; } });
Object.defineProperty(exports, "createModelServiceConfig", { enumerable: true, get: function () { return components_1.createModelServiceConfig; } });
var MODELS = __importStar(require("./constant"));
exports.MODELS = MODELS;
