// @antv/xflow-core/lib/model-service/model-registry.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModelRegistry = void 0;
var disposable_1 = require("../common/disposable");
var interface_1 = require("../xflow-main/interface");
var rx_model_1 = require("../common/rx-model");
var deferred_1 = require("../common/deferred");
var graph_provider_1 = require("../xflow-main/graph/graph-provider");
var interface_2 = require("./interface");
var mana_syringe_1 = require("mana-syringe");
var ModelRegistry = /** @class */ (function () {
    function ModelRegistry() {
        var _this = this;
        /** disposables */
        this.toDispose = new disposable_1.DisposableCollection();
        /** Deferred Model Map */
        this.deferredModelMap = new Map();
        /**
         * 注册model
         * @param options IModelOptions<T>
         */
        this.registerModel = function (options) {
            var id = options.id, getInitialValue = options.getInitialValue, modelFactory = options.modelFactory;
            var toDispose = new disposable_1.DisposableCollection();
            var defer = _this.ensureModel(id);
            if (defer.isResolved) {
                console.error(options, 'model has been registerd');
                return;
            }
            var initialValue = getInitialValue ? getInitialValue() : rx_model_1.NsModel.EMPTY_VALUE;
            var model = modelFactory
                ? modelFactory()
                : new rx_model_1.RxModel(initialValue);
            if (rx_model_1.NsModel.isValidValue(initialValue)) {
                defer.resolve(model);
            }
            if (options.watchChange) {
                /** 绑定watch事件 */
                options.watchChange(model, _this).then(function (d) {
                    /** createModel 后 resolve */
                    if (!defer.isResolved) {
                        defer.resolve(model);
                    }
                    _this.toDispose.pushAll([d, toDispose]);
                    toDispose.push(d);
                });
            }
            return toDispose;
        };
        /**
         * 查找 model
         * @param token: Token<T>
         */
        this.findDeferredModel = function (token) {
            return _this.deferredModelMap.get(token);
        };
        /**
         *  消费Model: await model resolve
         * @param token: Token<T>
         */
        this.awaitModel = function (token) {
            var defer = _this.ensureModel(token);
            return defer.promise;
        };
        /**
         * 注册 定义在IModelOptionProvider中的Model
         */
        this.registerRuntimeModel = function () { return __awaiter(_this, void 0, void 0, function () {
            var modelRegisterFunc, graphInstance;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.modelOptionProvider.getOption()];
                    case 1:
                        modelRegisterFunc = (_a.sent()).modelRegisterFunc;
                        return [4 /*yield*/, this.graphProvider.getGraphInstance()];
                    case 2:
                        graphInstance = _a.sent();
                        if (modelRegisterFunc) {
                            modelRegisterFunc(this, graphInstance);
                        }
                        return [2 /*return*/];
                }
            });
        }); };
        /**
         * 确保调用时有可用的Model
         * @param id ModelId
         */
        this.ensureModel = function (token) {
            var existDeferred = _this.deferredModelMap.get(token);
            if (existDeferred) {
                return existDeferred;
            }
            /** 注册ModelDeferred */
            var newDeferred = new deferred_1.Deferred();
            _this.deferredModelMap.set(token, newDeferred);
            _this.toDispose.push(disposable_1.Disposable.create(function () { return _this.deferredModelMap.delete(token); }));
            return newDeferred;
        };
    }
    /** app启动时，收集Model扩展点的注册项 */
    ModelRegistry.prototype.onStart = function () {
        var contributions = this.contributionProvider.getContributions();
        for (var _i = 0, contributions_1 = contributions; _i < contributions_1.length; _i++) {
            var contribution = contributions_1[_i];
            contribution.registerModel(this);
        }
        this.registerRuntimeModel();
    };
    /** app停止的逻辑 */
    ModelRegistry.prototype.onStop = function () {
        this.toDispose.dispose();
    };
    __decorate([
        (0, mana_syringe_1.contrib)(interface_2.IModelContribution),
        __metadata("design:type", Object)
    ], ModelRegistry.prototype, "contributionProvider", void 0);
    __decorate([
        (0, mana_syringe_1.inject)(interface_2.IModelOptionProvider),
        __metadata("design:type", Object)
    ], ModelRegistry.prototype, "modelOptionProvider", void 0);
    __decorate([
        (0, mana_syringe_1.inject)(graph_provider_1.IGraphProvider),
        __metadata("design:type", Object)
    ], ModelRegistry.prototype, "graphProvider", void 0);
    ModelRegistry = __decorate([
        (0, mana_syringe_1.singleton)({ contrib: [interface_1.IFrontendApplicationContribution, interface_2.IModelService] })
    ], ModelRegistry);
    return ModelRegistry;
}());
exports.ModelRegistry = ModelRegistry;
