// @antv/xflow-core/lib/command-contributions/command-contribution.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerXFlowCommandContribution = exports.XFlowCommandContribution = void 0;
/* eslint-disable @typescript-eslint/no-redeclare */
require("reflect-metadata");
/** Application 扩展依赖 */
var interface_1 = require("../command/interface");
var mana_syringe_1 = require("mana-syringe");
var interface_2 = require("../hooks/interface");
var disposable_1 = require("../common/disposable");
/** Commands */
var node_1 = require("./node");
var graph_1 = require("./graph");
var edge_1 = require("./edge");
var group_1 = require("./group");
var models_1 = require("./models");
/** Command Hooks*/
var xflow_hook_1 = require("@antv/xflow-hook");
var config_1 = require("./config");
/** Commands 配置项目*/
var hookhubList = __spreadArray(__spreadArray(__spreadArray(__spreadArray(__spreadArray([], node_1.hookhubList, true), edge_1.hookhubList, true), group_1.hookhubList, true), graph_1.hookhubList, true), models_1.hookhubList, true);
var defaultHookFactory = function () { return new xflow_hook_1.HookHub(); };
var XFlowCommandContribution = /** @class */ (function () {
    function XFlowCommandContribution() {
        var _this = this;
        /** 注册钩子 */
        this.registerHook = function (hooks) { return __awaiter(_this, void 0, void 0, function () {
            var d;
            var _this = this;
            return __generator(this, function (_a) {
                d = hooks.x6Events.registerHook({
                    name: 'bind group node move event',
                    handler: function (args) { return __awaiter(_this, void 0, void 0, function () {
                        var event;
                        var _this = this;
                        return __generator(this, function (_a) {
                            event = {
                                eventName: 'node:moving',
                                callback: function (_a) {
                                    var node = _a.node;
                                    return __awaiter(_this, void 0, void 0, function () {
                                        var isGroup, graph, groupId, group, isCollapsed, originSize, hasChange, originPosition, x, y, cornerX, cornerY, childs, groupData;
                                        return __generator(this, function (_b) {
                                            isGroup = node.prop('isGroup');
                                            if (isGroup) {
                                                node.prop('originPosition', node.getPosition());
                                                return [2 /*return*/];
                                            }
                                            graph = node.model.graph;
                                            groupId = node.getData().group;
                                            group = graph.getNodes().find(function (cell) { return cell.id === groupId; });
                                            if (!group) {
                                                return [2 /*return*/];
                                            }
                                            isCollapsed = group.getProp('isCollapsed');
                                            originSize = group.getProp('originSize');
                                            hasChange = false;
                                            if (originSize == null) {
                                                originSize = group.size();
                                                group.prop('originSize', originSize);
                                            }
                                            originPosition = group.prop('originPosition');
                                            if (originPosition == null) {
                                                originPosition = group.getPosition();
                                                group.prop('originPosition', originPosition);
                                            }
                                            x = originPosition.x;
                                            y = originPosition.y;
                                            cornerX = originPosition.x + originSize.width;
                                            cornerY = originPosition.y + originSize.height;
                                            childs = group.getChildren();
                                            if (childs) {
                                                childs.forEach(function (child) {
                                                    var bbox = child.getBBox().inflate(12);
                                                    var corner = bbox.getCorner();
                                                    if (bbox.x < x) {
                                                        x = bbox.x;
                                                        hasChange = true;
                                                    }
                                                    if (bbox.y < y) {
                                                        y = bbox.y;
                                                        hasChange = true;
                                                    }
                                                    if (corner.x > cornerX) {
                                                        cornerX = corner.x;
                                                        hasChange = true;
                                                    }
                                                    if (corner.y > cornerY) {
                                                        cornerY = corner.y;
                                                        hasChange = true;
                                                    }
                                                });
                                            }
                                            if (hasChange) {
                                                group.prop({
                                                    position: { x: x, y: y },
                                                    size: { width: cornerX - x, height: cornerY - y },
                                                });
                                                groupData = __assign(__assign({}, group.getData()), { x: x, y: y, width: cornerX - x, height: cornerY - y });
                                                if (isCollapsed !== true) {
                                                    groupData.groupChildrenSize = { width: cornerX - x, height: cornerY - y };
                                                }
                                                group.setData(groupData);
                                            }
                                            return [2 /*return*/];
                                        });
                                    });
                                },
                            };
                            args.push(event);
                            return [2 /*return*/];
                        });
                    }); },
                });
                return [2 /*return*/, disposable_1.Disposable.create(function () {
                        d.dispose();
                    })];
            });
        }); };
        /** 注册钩子 */
        this.registerHookHub = function (registry) { return __awaiter(_this, void 0, void 0, function () {
            var toDispose;
            return __generator(this, function (_a) {
                toDispose = new disposable_1.DisposableCollection();
                hookhubList.forEach(function (_a) {
                    var hookKey = _a.hookKey, _b = _a.createHook, createHook = _b === void 0 ? defaultHookFactory : _b;
                    var d = registry.registerHookHub(hookKey, createHook());
                    toDispose.push(d);
                });
                this.commandConfig.getConfig().then(function (_a) {
                    var getContributions = _a.getContributions;
                    var CommandContributions = getContributions();
                    CommandContributions.forEach(function (_a) {
                        var hookKey = _a.hookKey, _b = _a.createHook, createHook = _b === void 0 ? defaultHookFactory : _b;
                        var d = registry.registerHookHub(hookKey, createHook());
                        toDispose.push(d);
                    });
                });
                return [2 /*return*/, toDispose];
            });
        }); };
    }
    /** 注册画布节点命令 */
    XFlowCommandContribution.prototype.registerGraphCommands = function (registry) {
        var _this = this;
        /** 注册内置的命令 */
        hookhubList.forEach(function (_a) {
            var command = _a.command;
            registry.registerCommand(command, {
                createCommand: _this.commandFactory,
            });
        });
        /** 注册外部传入的命令 */
        this.commandConfig.getConfig().then(function (_a) {
            var getContributions = _a.getContributions;
            var CommandContributions = getContributions();
            CommandContributions.forEach(function (cmd) {
                registry.registerCommand(cmd.command, {
                    createCommand: _this.commandFactory,
                });
            });
        });
    };
    __decorate([
        (0, mana_syringe_1.inject)(interface_1.IGraphCommandFactory),
        __metadata("design:type", Function)
    ], XFlowCommandContribution.prototype, "commandFactory", void 0);
    __decorate([
        (0, mana_syringe_1.inject)(config_1.CommandConfig),
        __metadata("design:type", config_1.CommandConfig
        /** 注册画布节点命令 */
        )
    ], XFlowCommandContribution.prototype, "commandConfig", void 0);
    XFlowCommandContribution = __decorate([
        (0, mana_syringe_1.singleton)({ contrib: [interface_1.IGraphCommandContribution, interface_2.IHookContribution] })
    ], XFlowCommandContribution);
    return XFlowCommandContribution;
}());
exports.XFlowCommandContribution = XFlowCommandContribution;
var registerXFlowCommandContribution = function (register, commandConfig) {
    /** 扩展 用户自定义命令 */
    var configContributions = commandConfig.getCommandContributions();
    configContributions.forEach(function (execution) {
        var CommandHandler = execution.CommandHandler, command = execution.command;
        register({
            token: { token: interface_1.ICommandHandler, named: command.id },
            useClass: CommandHandler,
        });
    });
    /** 扩展 Graph 命令 */
    (0, graph_1.registerGraphCommand)(register);
    /** 扩展 Node 命令 */
    (0, node_1.registerNodeCommand)(register);
    /** 扩展 Edge 命令 */
    (0, edge_1.registerEdgeCommand)(register);
    /** 扩展 Edge 命令 */
    (0, group_1.registerGroupCommand)(register);
    /** 扩展 Observable 命令 */
    (0, models_1.registerModelServiceCommand)(register);
    /** 注册 XFlowCommandContribution */
    register(XFlowCommandContribution);
};
exports.registerXFlowCommandContribution = registerXFlowCommandContribution;
