// @antv/xflow-core/lib/command-contributions/mapping-service.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GraphMappingHelper = void 0;
var mana_syringe_1 = require("mana-syringe");
var GraphMappingHelper = /** @class */ (function () {
    function GraphMappingHelper() {
        var _this = this;
        this.groupNodeMap = new Map();
        // node 的 mapping
        this.nodeMap = new Map();
        this.nodeMappingRecord = new Map();
        // port 的 mapping
        this.portMap = new Map();
        this.portMappingRecord = new Map();
        // edge 的 mapping
        this.edgeMap = new Map();
        this.edgeMappingRecord = new Map();
        // 区分类型
        this.getNodesByType = function (nodes) {
            return nodes.reduce(function (acc, node) {
                if (node.isGroup) {
                    acc.groupNodes.push(node);
                }
                else {
                    acc.normalNodes.push(node);
                }
                return acc;
            }, { normalNodes: [], groupNodes: [] });
        };
        // 增加已有节点
        this.getNodePorts = function (node) {
            if (Array.isArray(node.ports)) {
                return node.ports;
            }
            if (node.ports && Array.isArray(node.ports.items)) {
                return node.ports.items;
            }
            return [];
        };
        // 增加已有节点
        this.addNodes = function (nodes) {
            nodes.forEach(function (node) {
                if (node.isGroup) {
                    _this.groupNodeMap.set(node.id, node);
                }
                else {
                    var ports = _this.getNodePorts(node);
                    _this.nodeMap.set(node.id, node);
                    ports.forEach(function (port) {
                        _this.portMap.set(port.id, port);
                    });
                }
            });
        };
        // 增加已有edge
        this.addEdge = function (edges) {
            edges.forEach(function (edge) {
                _this.edgeMap.set(edge.id, edge);
            });
        };
        // 添加node的映射关系
        this.buildNodeMapping = function (currentNode, nextNode) {
            var currentNodeId = currentNode.id || currentNode.originId;
            var nextNodeId = nextNode.id;
            _this.nodeMappingRecord.set(currentNodeId, nextNodeId);
            var ports = _this.getNodePorts(currentNode);
            ports.forEach(function (oldPort, index) {
                var newPort = nextNode.ports[index];
                if (newPort) {
                    _this.portMap.set(newPort.id, newPort);
                    _this.portMappingRecord.set(oldPort.id, newPort.id);
                }
            });
        };
        // 更新Edge的NodeId和PortId
        this.createEdgeBetweenNodes = function (edgeConfig) {
            var source = edgeConfig.source, sourcePortId = edgeConfig.sourcePortId, target = edgeConfig.target, targetPortId = edgeConfig.targetPortId;
            return __assign(__assign({}, edgeConfig), { source: _this.nodeMappingRecord.get(source), target: _this.nodeMappingRecord.get(target), sourcePortId: _this.portMappingRecord.get(sourcePortId), targetPortId: _this.portMappingRecord.get(targetPortId) });
        };
        // 更新GroupNode的NodeId
        this.buildGroupRelations = function (groupNode) {
            var groupChildren = groupNode.groupChildren;
            var newGroupChildren = groupChildren.map(function (id) {
                return _this.nodeMappingRecord.get(id);
            });
            return __assign(__assign({}, groupNode), { groupChildren: newGroupChildren });
        };
    }
    GraphMappingHelper = __decorate([
        (0, mana_syringe_1.injectable)()
    ], GraphMappingHelper);
    return GraphMappingHelper;
}());
exports.GraphMappingHelper = GraphMappingHelper;
