// @antv/xflow-core/es/command-contributions/node/node-move.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { inject, injectable } from 'mana-syringe';
import { ICommandHandler, ICommandContextProvider } from '../../command/interface';
import { HookHub } from '@antv/xflow-hook';
import { XFlowNodeCommands } from '../constant';
import { Disposable } from '../../common/disposable';
export var NsMoveNode;
(function (NsMoveNode) {
    /** Command: 用于注册named factory */
    NsMoveNode.command = XFlowNodeCommands.MOVE_NODE;
    /** hookName */
    NsMoveNode.hookKey = 'moveNode';
    /** 创建 hook */
    NsMoveNode.createHook = () => {
        return new HookHub();
    };
})(NsMoveNode || (NsMoveNode = {}));
let MoveNodeCommand = 
/** 创建节点命令 */
class MoveNodeCommand {
    constructor() {
        /** 执行Cmd */
        this.execute = () => __awaiter(this, void 0, void 0, function* () {
            const ctx = this.contextProvider();
            const { args, hooks: runtimeHook } = ctx.getArgs();
            const hooks = ctx.getHooks();
            const result = yield hooks.moveNode.call(args, (handlerArgs) => __awaiter(this, void 0, void 0, function* () {
                const { nodePositionService } = handlerArgs;
                /** nodePositionService 返回false时不更新position */
                if (nodePositionService) {
                    const canDel = yield nodePositionService(handlerArgs);
                    if (!canDel)
                        return { err: 'service rejected' };
                }
                const { dx = 0, dy = 0, x, y, duration = 150 } = handlerArgs.position;
                const x6Graph = yield ctx.getX6Graph();
                const node = x6Graph.getCellById(handlerArgs.id);
                if (node) {
                    let nextX = x;
                    let nextY = y;
                    const { x: preX, y: preY } = node.position();
                    let undo = () => {
                        node.position(preX, preY, { silent: false });
                    };
                    if (dx || dy) {
                        nextX = dx + preX;
                        nextY = dy + preY;
                        node.translate(dx, dy, { transition: { duration } });
                        undo = () => node.translate(-dx, -dy, { transition: { duration } });
                    }
                    else {
                        node.position(nextX, nextY, { silent: false });
                    }
                    /** add undo  */
                    ctx.addUndo(Disposable.create(() => {
                        undo();
                    }));
                    return { err: null, nextX, nextY };
                }
            }), runtimeHook);
            ctx.setResult(result);
            return this;
        });
        /** undo cmd */
        this.undo = () => __awaiter(this, void 0, void 0, function* () {
            const ctx = this.contextProvider();
            if (this.isUndoable()) {
                ctx.undo();
            }
            return this;
        });
        /** redo cmd */
        this.redo = () => __awaiter(this, void 0, void 0, function* () {
            if (!this.isUndoable()) {
                yield this.execute();
            }
            return this;
        });
    }
    isUndoable() {
        const ctx = this.contextProvider();
        return ctx.isUndoable();
    }
};
__decorate([
    inject(ICommandContextProvider),
    __metadata("design:type", Object)
], MoveNodeCommand.prototype, "contextProvider", void 0);
MoveNodeCommand = __decorate([
    injectable({
        token: { token: ICommandHandler, named: NsMoveNode.command.id },
    })
    /** 创建节点命令 */
], MoveNodeCommand);
export { MoveNodeCommand };
