// @antv/xflow-core/lib/xflow-main/components/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExtensionRegistry = exports.XFlowAppExtension = exports.XFlowCanvas = exports.useXflowPrefixCls = exports.useXFlowApp = exports.XFlow = void 0;
var react_1 = __importDefault(require("react"));
var classnames_1 = __importDefault(require("classnames"));
/** app */
var application_module_1 = require("../application-module");
var app_context_1 = require("./app-context");
Object.defineProperty(exports, "useXFlowApp", { enumerable: true, get: function () { return app_context_1.useXFlowApp; } });
/** app-extension */
var extension_context_1 = require("./extension-context");
var global_config_context_1 = require("./global-config-context");
Object.defineProperty(exports, "useXflowPrefixCls", { enumerable: true, get: function () { return global_config_context_1.useXflowPrefixCls; } });
var app_extension_module_1 = require("./app-extension-module");
Object.defineProperty(exports, "XFlowAppExtension", { enumerable: true, get: function () { return app_extension_module_1.XFlowAppExtensionMoudle; } });
var extension_registry_1 = require("./extension-registry");
Object.defineProperty(exports, "ExtensionRegistry", { enumerable: true, get: function () { return extension_registry_1.ExtensionRegistry; } });
/** graph */
var canvas_1 = require("./canvas");
Object.defineProperty(exports, "XFlowCanvas", { enumerable: true, get: function () { return canvas_1.XFlowCanvas; } });
/** command */
var model_service_1 = require("../../model-service");
var command_contributions_1 = require("../../command-contributions");
/** hook */
var hooks_1 = require("../../hooks");
/** UI model */
var toolbar_1 = require("../../toolbar");
var menu_1 = require("../../menu");
var XFlow = function (props) {
    var meta = props.meta, graphConfig = props.graphConfig, graphData = props.graphData, graphLayout = props.graphLayout, onLoad = props.onLoad, isAutoCenter = props.isAutoCenter, hookConfig = props.hookConfig, modelServiceConfig = props.modelServiceConfig, commandConfig = props.commandConfig, onAppConfigReady = props.onAppConfigReady, onAppDestroy = props.onAppDestroy, _a = props.children, children = _a === void 0 ? [] : _a, className = props.className;
    var _b = react_1.default.useState(), appRef = _b[0], setAppRef = _b[1];
    /** XFlow App 配置中心 */
    var extensionRegistry = (0, extension_registry_1.createExtensionRegistry)();
    /** didmount */
    react_1.default.useEffect(function () {
        /** before app start */
        if (onAppConfigReady) {
            onAppConfigReady(extensionRegistry);
        }
        /** 初始化应用 */
        var app = (0, application_module_1.initApp)(extensionRegistry);
        app.start().then(function () { return __awaiter(void 0, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        /** 保留引用 */
                        setAppRef(app);
                        if (!meta) return [3 /*break*/, 2];
                        return [4 /*yield*/, app.commandService.executeCommand(command_contributions_1.XFlowGraphCommands.LOAD_META.id, { meta: meta })];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2:
                        /** after app start */
                        if (onLoad) {
                            onLoad(app, extensionRegistry);
                        }
                        if (!graphData) return [3 /*break*/, 6];
                        if (!graphLayout) return [3 /*break*/, 4];
                        return [4 /*yield*/, app.commandService.executeCommand(command_contributions_1.XFlowGraphCommands.GRAPH_LAYOUT.id, __assign({ graphData: graphData }, graphLayout))];
                    case 3:
                        _a.sent();
                        _a.label = 4;
                    case 4: return [4 /*yield*/, app.commandService.executeCommand(command_contributions_1.XFlowGraphCommands.GRAPH_RENDER.id, { graphData: graphData })];
                    case 5:
                        _a.sent();
                        _a.label = 6;
                    case 6: return [2 /*return*/];
                }
            });
        }); });
        /** unmount */
        var destroy = function () { return __awaiter(void 0, void 0, void 0, function () {
            return __generator(this, function (_a) {
                onAppDestroy && onAppDestroy(app);
                return [2 /*return*/];
            });
        }); };
        return function () {
            destroy();
        };
        /** 不要删 保证只生成一次 */
        // eslint-disable-next-line
    }, []);
    /** 自动更新meta */
    react_1.default.useEffect(function () {
        if (appRef) {
            appRef.commandService.executeCommand(command_contributions_1.XFlowGraphCommands.LOAD_META.id, { meta: meta });
        }
        /** 不要删 只和meta联动 */
        // eslint-disable-next-line
    }, [meta]);
    /** 自动渲染画布内容 */
    react_1.default.useEffect(function () {
        var fn = function () { return __awaiter(void 0, void 0, void 0, function () {
            var x6Graph;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!appRef) return [3 /*break*/, 5];
                        if (!(graphData && graphLayout)) return [3 /*break*/, 2];
                        return [4 /*yield*/, appRef.commandService.executeCommand(command_contributions_1.XFlowGraphCommands.GRAPH_LAYOUT.id, __assign({ graphData: graphData }, graphLayout))];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2: return [4 /*yield*/, appRef.commandService.executeCommand(command_contributions_1.XFlowGraphCommands.GRAPH_RENDER.id, {
                            graphData: graphData,
                        })
                        /** 自动居中画布内容 */
                    ];
                    case 3:
                        _a.sent();
                        if (!isAutoCenter) return [3 /*break*/, 5];
                        return [4 /*yield*/, appRef.getGraphInstance()];
                    case 4:
                        x6Graph = _a.sent();
                        x6Graph.centerContent();
                        _a.label = 5;
                    case 5: return [2 /*return*/];
                }
            });
        }); };
        fn();
        /** 不要删 只和graphData联动 */
        // eslint-disable-next-line
    }, [graphData]);
    /** classname */
    var appClzName = classnames_1.default.apply(void 0, __spreadArray(['xflow-app-workspace',
        className], extensionRegistry.getContainerClassNames(), false));
    /** 判断是否需要自动渲染画布组件, 坐标相对于xflow-graph-root */
    var hasCanvasComponent = (Array.isArray(children) ? children : [children]).some(function (child) {
        return child && child.props.isXFlowCanvas;
    });
    return (react_1.default.createElement(app_context_1.XFlowAppInternalProvider, { app: appRef },
        react_1.default.createElement(extension_context_1.ExtensionRegistryContext.Provider, { value: extensionRegistry },
            react_1.default.createElement("div", { className: appClzName, id: extensionRegistry.getInstaceId() },
                !hasCanvasComponent && (react_1.default.createElement(canvas_1.XFlowCanvas, { config: graphConfig, position: { top: 0, bottom: 0, left: 0, right: 0 } })),
                children,
                react_1.default.createElement(model_service_1.ModelServiceRegistry, { config: modelServiceConfig }),
                react_1.default.createElement(command_contributions_1.CommandsRegistry, { config: commandConfig }),
                react_1.default.createElement(hooks_1.HookRegistry, { config: hookConfig }),
                react_1.default.createElement(toolbar_1.ToolbarRegistry, null),
                react_1.default.createElement(menu_1.MenuRegistry, null)))));
};
exports.XFlow = XFlow;
exports.XFlow.defaultProps = {};
exports.default = exports.XFlow;
