// @antv/xflow-core/lib/xflow-main/graph/config.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createGraphConfig = exports.registerGraphConfig = exports.IGraphOptionProvider = exports.GraphConfig = exports.NsGraphConfig = void 0;
var react_1 = __importDefault(require("react"));
var mana_syringe_1 = require("mana-syringe");
var deferred_1 = require("../../common/deferred");
var constants_1 = require("../../constants");
var uuid_1 = require("../../common/uuid");
var react_group_node_1 = require("./react-group-node");
var react_node_1 = require("./react-node");
var NsGraphConfig;
(function (NsGraphConfig) {
    NsGraphConfig.CONFIG_TYPE = 'GraphConfig';
    /** 默认的Node解析函数 */
    NsGraphConfig.defaultNodeTypeParser = function (nodeConfig) { return nodeConfig.renderKey; };
    /** 构造的Edge解析函数 */
    NsGraphConfig.defaultEdgeTypeParser = function (edgeConfig) { return edgeConfig.renderKey; };
    /** XFlow默认的React组件 */
    NsGraphConfig.defaultNodeMapValue = [
        [constants_1.XFLOW_DEFAULT_NODE, react_node_1.XFlowDefaultNode],
        [constants_1.XFLOW_DEFAULT_GROUP_NODE, react_group_node_1.XFlowDefaultGroupNode],
    ];
})(NsGraphConfig = exports.NsGraphConfig || (exports.NsGraphConfig = {}));
var GraphConfig = /** @class */ (function () {
    /** 构造函数 */
    function GraphConfig() {
        var _this = this;
        /** 配置类型 */
        this.CONFIG_TYPE = NsGraphConfig.CONFIG_TYPE;
        /** 画布的属性 */
        this.x6Options = new deferred_1.Deferred();
        /** 画布的root节点 */
        this.rootContainer = new deferred_1.Deferred();
        /** 画布的dom节点 */
        this.graphContainer = new deferred_1.Deferred();
        /** 自定义节点 */
        this.nodeRender = new Map(NsGraphConfig.defaultNodeMapValue);
        /** 自定义边 */
        this.edgeRender = new Map();
        /** 解析node渲染类型的parser */
        this.nodeTypeParser = NsGraphConfig.defaultNodeTypeParser;
        /** 解析edge渲染类型的parser  */
        this.edgeTypeParser = NsGraphConfig.defaultEdgeTypeParser;
        /** 自定义事件 */
        this.events = [];
        this.mergeX6Config = function (config) { return __awaiter(_this, void 0, void 0, function () {
            var current;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getX6Config()];
                    case 1:
                        current = _a.sent();
                        return [2 /*return*/, Object.assign(current, config)];
                }
            });
        }); };
        this.getX6Config = function () { return __awaiter(_this, void 0, void 0, function () {
            var config;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.x6Options.promise];
                    case 1:
                        config = _a.sent();
                        return [2 /*return*/, config];
                }
            });
        }); };
        this.setX6Config = function (options) {
            var _a;
            var defaultOptions = _this.getDefaultGraphOptions();
            (_a = _this.x6Options) === null || _a === void 0 ? void 0 : _a.resolve(__assign(__assign({}, defaultOptions), options));
        };
        this.setRootContainer = function (ele) {
            ele && (_this === null || _this === void 0 ? void 0 : _this.rootContainer.resolve(ele));
        };
        this.setGraphContainer = function (ele) {
            var _a;
            ele && ((_a = _this.graphContainer) === null || _a === void 0 ? void 0 : _a.resolve(ele));
        };
        this.setEvents = function (events) {
            _this.events = events;
        };
        this.setDefaultNodeRender = function (component) {
            _this.nodeRender.set(constants_1.XFLOW_DEFAULT_NODE, component);
        };
        this.setDefaultEdgeRender = function (component) {
            _this.edgeRender.set(constants_1.XFLOW_DEFAULT_EDGE, component);
        };
        this.setNodeRender = function (renderKey, component) {
            _this.nodeRender.set(renderKey, component);
        };
        this.setEdgeRender = function (renderKey, component) {
            _this.edgeRender.set(renderKey, component);
        };
        /** 设置解析node render key的 parser */
        this.setNodeTypeParser = function (parser) {
            _this.nodeTypeParser = parser;
        };
        /** 设置解析edge render key的 parser */
        this.setEdgeTypeParser = function (parser) {
            _this.edgeTypeParser = parser;
        };
        /** 获取Graph参数 */
        this.getConfig = function () { return __awaiter(_this, void 0, void 0, function () {
            var _a, rootContainer, graphContainer, x6Options, graphConfig;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0: return [4 /*yield*/, Promise.all([
                            this.rootContainer.promise,
                            this.graphContainer.promise,
                            this.x6Options.promise,
                        ])];
                    case 1:
                        _a = _b.sent(), rootContainer = _a[0], graphContainer = _a[1], x6Options = _a[2];
                        graphConfig = {
                            xflowInstanceId: this.xflowInstanceId,
                            graphId: this.graphId,
                            nodeViewId: this.graphId,
                            rootContainer: rootContainer,
                            graphContainer: graphContainer,
                            x6Options: x6Options,
                            events: this.events,
                            nodeRender: this.nodeRender,
                            edgeRender: this.edgeRender,
                            nodeTypeParser: this.nodeTypeParser,
                            edgeTypeParser: this.edgeTypeParser,
                        };
                        return [2 /*return*/, graphConfig];
                }
            });
        }); };
        this.getDefaultGraphOptions = function () {
            var defaultOptions = {
                // 节点是否可旋转
                rotating: false,
                // 节点是否可调整大小
                resizing: false,
                // 节点连线规则配置（详细文档：https://X6.antv.vision/zh/docs/api/graph/interaction#connecting）
                connecting: {
                    snap: true,
                    dangling: false,
                    highlight: false,
                    connectionPoint: 'rect',
                    router: { name: 'er' },
                    connector: {
                        name: 'rounded',
                        args: {
                            radius: 15,
                        },
                    },
                },
                // 画布背景，支持颜色/图片/水印等（详细文档：https://X6.antv.vision/zh/docs/tutorial/basic/background）
                background: {},
                // 网格配置（详细文档：https://X6.antv.vision/zh/docs/tutorial/basic/grid）
                grid: {
                    visible: true,
                },
                // 点选/框选配置（详细文档：https://X6.antv.vision/zh/docs/tutorial/basic/selection）
                selecting: {
                    strict: true,
                    enabled: true,
                    multiple: true,
                    selectCellOnMoved: true,
                    showNodeSelectionBox: false,
                    // 框选可以选中edge
                    rubberEdge: true,
                    // 框选可以选中node
                    rubberNode: true,
                    movable: true,
                },
                // 对齐线配置，辅助移动节点排版（详细文档：https://X6.antv.vision/zh/docs/tutorial/basic/snapline）
                snapline: {
                    enabled: false,
                },
                // 撤销/重做能力（详细文档：https://X6.antv.vision/zh/docs/tutorial/basic/history）
                history: {
                    enabled: false,
                },
                // 剪切板，支持跨画布的复制/粘贴（详细文档：https://X6.antv.vision/zh/docs/tutorial/basic/clipboard）
                clipboard: {
                    enabled: true,
                    useLocalStorage: true,
                },
                // 使画布具备滚动、平移、居中、缩放等能力（详细文档：https://X6.antv.vision/zh/docs/tutorial/basic/scroller）
                scroller: {
                    enabled: false,
                },
                // 滚轮缩放 （详细文档：https://X6.antv.vision/zh/docs/tutorial/basic/mousewheel）
                mousewheel: {
                    enabled: true,
                    minScale: constants_1.MIN_ZOOM,
                    maxScale: constants_1.MAX_ZOOM,
                    factor: 1.1,
                    modifiers: ['ctrl', 'meta'],
                },
                /** 普通画布, 支持拖拽平移 */
                panning: {
                    enabled: true,
                },
                /** 缩放参数 */
                scaling: {
                    min: constants_1.MIN_ZOOM,
                    max: constants_1.MAX_ZOOM,
                },
                /** keyboad */
                keyboard: {
                    enabled: true,
                },
                /** 定制节点和边的交互行为 */
                interacting: {
                    /** 节点默认可以被移动 */
                    nodeMovable: true,
                    /** 边上标签默认不可以被移动 */
                    edgeLabelMovable: false,
                },
                async: false,
            };
            return defaultOptions;
        };
        this.dispose = function () {
            // TODO: 统一看下defer的销毁
            _this.rootContainer = new deferred_1.Deferred();
            _this.graphContainer = new deferred_1.Deferred();
            _this.x6Options = new deferred_1.Deferred();
        };
        this.graphId = (0, uuid_1.uuidv4)();
    }
    return GraphConfig;
}());
exports.GraphConfig = GraphConfig;
exports.IGraphOptionProvider = Symbol('IGraphOptionProvider');
var registerGraphConfig = function (register, graphConfig) {
    register(exports.IGraphOptionProvider, {
        lifecycle: mana_syringe_1.Syringe.Lifecycle.singleton,
        useDynamic: function () {
            return {
                getOptions: function () { return __awaiter(void 0, void 0, void 0, function () {
                    return __generator(this, function (_a) {
                        return [2 /*return*/, graphConfig.getConfig()];
                    });
                }); },
            };
        },
    });
};
exports.registerGraphConfig = registerGraphConfig;
var createGraphConfig = function (addOptions) {
    return function (props) {
        /** bridge config and props */
        var propsContainer = react_1.default.useMemo(function () { return ({ getValue: function () { return ({}); } }); }, []);
        propsContainer.getValue = function () { return props; };
        var graphConfig = react_1.default.useMemo(function () {
            var config = new GraphConfig();
            addOptions(config, propsContainer);
            return config;
        }, [propsContainer]);
        return graphConfig;
    };
};
exports.createGraphConfig = createGraphConfig;
