// @antv/xflow-core/lib/xflow-main/graph/graph-provider.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerGraphModule = exports.GraphManager = exports.IModelServiceProvider = exports.ICommandServiceProvider = exports.IGraphProvider = void 0;
var react_dom_1 = __importDefault(require("react-dom"));
var x6_1 = require("@antv/x6");
var disposable_1 = require("../../common/disposable");
var deferred_1 = require("../../common/deferred");
var config_1 = require("./config");
var mana_syringe_1 = require("mana-syringe");
var command_1 = require("../../command");
var interface_1 = require("../../model-service/interface");
var hooks_1 = require("../../hooks");
var context_1 = require("../../command-contributions/components/context");
var constants_1 = require("../../constants");
/** !!! 所有需要使用Graph相关信息的地方, 都统一使用IGraphProvider来获取 */
exports.IGraphProvider = Symbol('IGraphProvider');
exports.ICommandServiceProvider = Symbol('ICommandServiceProvider');
exports.IModelServiceProvider = Symbol('IModelServiceProvider');
var GraphManager = /** @class */ (function () {
    function GraphManager() {
        var _this = this;
        /** 处理画布实例的销毁 */
        this.toDisposce = new disposable_1.DisposableCollection();
        /** 储存画布实例 */
        this.graphMap = new Map();
        /** 获取X6 Graph 实例 */
        this.getGraph = function (graphId) { return __awaiter(_this, void 0, void 0, function () {
            var graphDefer, options, hooks_2, mergedOptions, graphContainer, edgeRender, nodeRender, edgeTypeParser_1, edgeRenderMap_1, clientHeight, clientWidth, commandService_1, modelService_1, graph_1;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        graphDefer = this.graphMap.get(graphId);
                        if (!!graphDefer) return [3 /*break*/, 8];
                        graphDefer = new deferred_1.Deferred();
                        this.graphMap.set(graphId, graphDefer);
                        return [4 /*yield*/, this.optionProvider.getOptions()
                            /** 获取hooks实例： hooks方便其他组件可以修改注册graphOpions/events */
                        ];
                    case 1:
                        options = _a.sent();
                        hooks_2 = this.hookService.hookProvider();
                        return [4 /*yield*/, hooks_2.graphOptions.call(options.x6Options)];
                    case 2:
                        mergedOptions = _a.sent();
                        graphContainer = options.graphContainer, edgeRender = options.edgeRender, nodeRender = options.nodeRender, edgeTypeParser_1 = options.edgeTypeParser;
                        return [4 /*yield*/, hooks_2.reactEdgeLabelRender.call(edgeRender)
                            /** 执行hooks：更新nodeRender */
                        ];
                    case 3:
                        edgeRenderMap_1 = _a.sent();
                        /** 执行hooks：更新nodeRender */
                        return [4 /*yield*/, hooks_2.reactNodeRender.call(nodeRender)];
                    case 4:
                        /** 执行hooks：更新nodeRender */
                        _a.sent();
                        clientHeight = graphContainer.clientHeight, clientWidth = graphContainer.clientWidth;
                        return [4 /*yield*/, this.commandServiceProvider.getCommandService()];
                    case 5:
                        commandService_1 = _a.sent();
                        return [4 /*yield*/, this.modelServiceProvider.getModelService()
                            /** 实例化 X6 Graph */
                        ];
                    case 6:
                        modelService_1 = _a.sent();
                        graph_1 = new x6_1.Graph(__assign(__assign({ container: graphContainer, width: clientWidth, height: clientHeight }, mergedOptions), { 
                            /** X6提供了边渲染的钩子, 可以在这里设置边上需要渲染的React内容 */
                            onEdgeLabelRendered: function (args) {
                                var edge = args.edge, container = args.container;
                                _this.renderEdgeReactLabel({
                                    edge: edge,
                                    container: container,
                                    edgeRenderMap: edgeRenderMap_1,
                                    edgeTypeParser: edgeTypeParser_1,
                                    modelService: modelService_1,
                                    commandService: commandService_1,
                                });
                            } }));
                        /** 执行hooks：绑定事件执行api */
                        return [4 /*yield*/, hooks_2.afterGraphInit.call({
                                graph: graph_1,
                                commandService: commandService_1,
                                modelService: modelService_1,
                            })];
                    case 7:
                        /** 执行hooks：绑定事件执行api */
                        _a.sent();
                        graphDefer.resolve(graph_1);
                        graph_1.on('node:moved', function (_a) {
                            var node = _a.node;
                            var nodeData = node.getData();
                            var position = node.position();
                            node.setData(__assign(__assign({}, nodeData), { x: position === null || position === void 0 ? void 0 : position.x, y: position === null || position === void 0 ? void 0 : position.y }));
                        });
                        graph_1.on('node:resized', function (_a) {
                            var node = _a.node;
                            var nodeData = node.getData();
                            var size = node.size();
                            node.setData(__assign(__assign({}, nodeData), { width: size === null || size === void 0 ? void 0 : size.width, height: size === null || size === void 0 ? void 0 : size.height }));
                        });
                        this.toDisposce.push(disposable_1.Disposable.create(function () { return __awaiter(_this, void 0, void 0, function () {
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0: return [4 /*yield*/, hooks_2.beforeGraphDestory.call({
                                            graph: graph_1,
                                            commandService: commandService_1,
                                            modelService: modelService_1,
                                        })];
                                    case 1:
                                        _a.sent();
                                        this.graphMap.delete(graphId);
                                        graph_1.dispose();
                                        return [2 /*return*/];
                                }
                            });
                        }); }));
                        _a.label = 8;
                    case 8: return [2 /*return*/, graphDefer.promise];
                }
            });
        }); };
        this.renderEdgeReactLabel = function (args) { return __awaiter(_this, void 0, void 0, function () {
            var container, edgeTypeParser, edge, edgeRenderMap, commandService, modelService, renderKey, reactComponent, content, WrappedReactComponent;
            return __generator(this, function (_a) {
                container = args.container, edgeTypeParser = args.edgeTypeParser, edge = args.edge, edgeRenderMap = args.edgeRenderMap, commandService = args.commandService, modelService = args.modelService;
                renderKey = edgeTypeParser(edge === null || edge === void 0 ? void 0 : edge.data) || constants_1.XFLOW_DEFAULT_EDGE;
                reactComponent = edgeRenderMap.get(renderKey);
                if (!reactComponent) {
                    return [2 /*return*/];
                }
                if (edge && (edge === null || edge === void 0 ? void 0 : edge.data) && reactComponent) {
                    content = this.edgeAppendForeignObject(edge, container);
                    WrappedReactComponent = (0, context_1.getEdgeReactComponent)(reactComponent, commandService, modelService)(edge);
                    react_dom_1.default.render(WrappedReactComponent, content);
                }
                return [2 /*return*/];
            });
        }); };
    }
    /** 实现在连线上渲染React节点 */
    GraphManager.prototype.edgeAppendForeignObject = function (x6Edge, container) {
        var fo = x6_1.Dom.createSvgElement('foreignObject');
        var body = x6_1.Dom.createElementNS('body', x6_1.Dom.ns.xhtml);
        var content = x6_1.Dom.createElementNS('div', x6_1.Dom.ns.xhtml);
        var edgeData = x6Edge === null || x6Edge === void 0 ? void 0 : x6Edge.data;
        var foWdith = (edgeData === null || edgeData === void 0 ? void 0 : edgeData.edgeContentWidth) || 100;
        var foHeight = (edgeData === null || edgeData === void 0 ? void 0 : edgeData.edgeContentHeigt) || 30;
        fo.setAttribute('width', "" + foWdith);
        fo.setAttribute('height', "" + foHeight);
        fo.setAttribute('x', "" + (-1.0 * foWdith) / 2);
        fo.setAttribute('y', "" + (-1.0 * foHeight) / 2);
        body.setAttribute('xhtmls', x6_1.Dom.ns.xhtml);
        body.style.width = '100%';
        body.style.height = '100%';
        body.style.padding = '0';
        body.style.margin = '0';
        body.style.background = 'unset';
        body.style.overflow = 'visible';
        body.className = 'xflow-edge-label-body';
        content.style.width = '100%';
        content.style.height = '100%';
        body.appendChild(content);
        fo.appendChild(body);
        container.appendChild(fo);
        return content;
    };
    GraphManager.prototype.dispose = function () {
        this.toDisposce.dispose();
    };
    __decorate([
        (0, mana_syringe_1.inject)(config_1.IGraphOptionProvider),
        __metadata("design:type", Object)
    ], GraphManager.prototype, "optionProvider", void 0);
    __decorate([
        (0, mana_syringe_1.inject)(hooks_1.IHookService),
        __metadata("design:type", Object)
    ], GraphManager.prototype, "hookService", void 0);
    __decorate([
        (0, mana_syringe_1.inject)(exports.ICommandServiceProvider),
        __metadata("design:type", Object)
    ], GraphManager.prototype, "commandServiceProvider", void 0);
    __decorate([
        (0, mana_syringe_1.inject)(exports.IModelServiceProvider),
        __metadata("design:type", Object)
    ], GraphManager.prototype, "modelServiceProvider", void 0);
    GraphManager = __decorate([
        (0, mana_syringe_1.singleton)()
    ], GraphManager);
    return GraphManager;
}());
exports.GraphManager = GraphManager;
var registerGraphModule = function (register) {
    /** 注册GraphManager */
    register(GraphManager);
    register(exports.IGraphProvider, {
        lifecycle: mana_syringe_1.Syringe.Lifecycle.singleton,
        useDynamic: function (context) {
            return {
                getGraphInstance: function () { return __awaiter(void 0, void 0, void 0, function () {
                    var graphManager, graphOptionProvider, graphConfig, x6Graph;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                graphManager = context.container.get(GraphManager);
                                graphOptionProvider = context.container.get(config_1.IGraphOptionProvider);
                                return [4 /*yield*/, (graphOptionProvider === null || graphOptionProvider === void 0 ? void 0 : graphOptionProvider.getOptions())];
                            case 1:
                                graphConfig = _a.sent();
                                return [4 /*yield*/, (graphManager === null || graphManager === void 0 ? void 0 : graphManager.getGraph(graphConfig.graphId))];
                            case 2:
                                x6Graph = _a.sent();
                                return [2 /*return*/, x6Graph];
                        }
                    });
                }); },
                getGraphOptions: function () { return __awaiter(void 0, void 0, void 0, function () {
                    var graphOptionProvider;
                    return __generator(this, function (_a) {
                        graphOptionProvider = context.container.get(config_1.IGraphOptionProvider);
                        return [2 /*return*/, graphOptionProvider === null || graphOptionProvider === void 0 ? void 0 : graphOptionProvider.getOptions()];
                    });
                }); },
            };
        },
    });
    register(exports.ICommandServiceProvider, {
        lifecycle: mana_syringe_1.Syringe.Lifecycle.singleton,
        useDynamic: function (context) {
            return {
                getCommandService: function () { return __awaiter(void 0, void 0, void 0, function () {
                    var commandService;
                    return __generator(this, function (_a) {
                        commandService = context.container.get(command_1.IGraphCommandService);
                        return [2 /*return*/, commandService];
                    });
                }); },
            };
        },
    });
    register(exports.IModelServiceProvider, {
        lifecycle: mana_syringe_1.Syringe.Lifecycle.singleton,
        useDynamic: function (context) {
            return {
                getModelService: function () { return __awaiter(void 0, void 0, void 0, function () {
                    var modelService;
                    return __generator(this, function (_a) {
                        modelService = context.container.get(interface_1.IModelService);
                        return [2 /*return*/, modelService];
                    });
                }); },
            };
        },
    });
};
exports.registerGraphModule = registerGraphModule;
