// @antv/xflow-core/es/xflow-main/application.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
/* eslint-disable @typescript-eslint/no-redeclare */
/* eslint-disable no-await-in-loop */
/* eslint-disable max-classes-per-file */
import { inject, singleton, contrib, Contribution } from 'mana-syringe';
import { Node as X6Node, Edge as X6Edge } from '@antv/x6';
import { IGraphProvider } from '../xflow-main/graph/graph-provider';
import { IGraphCommandService } from '../command/interface';
import { IModelService } from '../model-service';
import { IFrontendApplicationContribution } from './interface';
import { cellsToJson } from '../common/graph-utils';
export { IFrontendApplicationContribution } from './interface';
const TIMER_WARNING_THRESHOLD = 100;
let FrontendApplication = class FrontendApplication {
    constructor() {
        /** 获取画布配置项 */
        this.getGraphData = () => __awaiter(this, void 0, void 0, function* () {
            const graph = yield this.graphProvider.getGraphInstance();
            const cells = graph.getCells();
            return cellsToJson(cells);
        });
    }
    /** 启动app */
    start() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.startContributions();
            this.registerEventListeners();
        });
    }
    /** 获取画布实例 */
    getGraphInstance() {
        return this.graphProvider.getGraphInstance();
    }
    /** 获取画布配置项 */
    getGraphConfig() {
        return this.graphProvider.getGraphOptions();
    }
    /** 获取画布所有节点 */
    getAllNodes() {
        return __awaiter(this, void 0, void 0, function* () {
            const graph = yield this.graphProvider.getGraphInstance();
            return graph.getNodes();
        });
    }
    /** 获取画布节点 */
    getNodeById(nodeId) {
        return __awaiter(this, void 0, void 0, function* () {
            const graph = yield this.graphProvider.getGraphInstance();
            return graph.getCellById(nodeId);
        });
    }
    /** 获取画布所有连线 */
    getAllEdges() {
        return __awaiter(this, void 0, void 0, function* () {
            const graph = yield this.graphProvider.getGraphInstance();
            return graph.getEdges();
        });
    }
    /** 获取画布连线 */
    getEdgeById(edgeId) {
        return __awaiter(this, void 0, void 0, function* () {
            const graph = yield this.graphProvider.getGraphInstance();
            return graph.getCellById(edgeId);
        });
    }
    /** 更新节点样式 */
    updateNodeAttrs(node, attrs) {
        return __awaiter(this, void 0, void 0, function* () {
            if (node instanceof X6Node) {
                node.setAttrs(attrs);
            }
            else {
                const x6Node = yield this.getNodeById(node);
                x6Node.setAttrs(attrs);
            }
        });
    }
    /** 更新连线样式 */
    updateEdgeAttrs(edge, attrs) {
        return __awaiter(this, void 0, void 0, function* () {
            if (edge instanceof X6Edge) {
                edge.setAttrs(attrs);
            }
            else {
                const x6Edge = yield this.getEdgeById(edge);
                x6Edge.setAttrs(attrs);
            }
        });
    }
    /** 平移画布 */
    translateGraph(tx, ty) {
        return __awaiter(this, void 0, void 0, function* () {
            const graph = yield this.graphProvider.getGraphInstance();
            const currentTranslate = graph.translate();
            graph.translate(currentTranslate.tx + tx, currentTranslate.ty + ty);
        });
    }
    /** 暴露命令的执行接口 */
    executeCommand(commandId, cmdArgs, hook = []) {
        return this.commandService.executeCommand(commandId, cmdArgs, hook);
    }
    /** 暴露命令的批量执行接口 */
    executeCommandPipeline(cmdOptions) {
        return this.commandService.executeCommandPipeline(cmdOptions);
    }
    /**
     * Register global event listeners.
     */
    registerEventListeners() {
        /** 触发app的卸载逻辑 */
        window.addEventListener('unload', () => {
            this.stopContributions();
        });
    }
    /**
     * Initialize and start the frontend application contributions.
     */
    startContributions() {
        return __awaiter(this, void 0, void 0, function* () {
            console.log(this.contributions.getContributions());
            for (const contribution of this.contributions.getContributions()) {
                if (contribution.onStart) {
                    try {
                        yield this.measure(`${contribution.constructor.name}.onStart`, () => contribution.onStart(this));
                    }
                    catch (error) {
                        console.error('Could not start contribution', error);
                    }
                }
            }
        });
    }
    /**
     * Stop the frontend application contributions. This is called when the window is unloaded.
     */
    stopContributions() {
        console.info('>>> Stopping frontend contributions...');
        for (const contribution of this.contributions.getContributions()) {
            if (contribution.onStop) {
                try {
                    contribution.onStop(this);
                }
                catch (error) {
                    console.error('Could not stop contribution', error);
                }
            }
        }
        console.info('<<< All frontend contributions have been stopped.');
    }
    measure(name, fn) {
        return __awaiter(this, void 0, void 0, function* () {
            const startMark = `${name}-start`;
            const endMark = `${name}-end`;
            performance.mark(startMark);
            const result = yield fn();
            performance.mark(endMark);
            performance.measure(name, startMark, endMark);
            for (const item of performance.getEntriesByName(name)) {
                const contribution = `Frontend ${item.name}`;
                if (item.duration > TIMER_WARNING_THRESHOLD) {
                    console.warn(`${contribution} is slow, took: ${item.duration.toFixed(1)} ms`);
                }
                else {
                    console.debug(`${contribution} took: ${item.duration.toFixed(1)} ms`);
                }
            }
            performance.clearMeasures(name);
            return result;
        });
    }
};
__decorate([
    contrib(IFrontendApplicationContribution),
    __metadata("design:type", Object)
], FrontendApplication.prototype, "contributions", void 0);
__decorate([
    inject(IGraphProvider),
    __metadata("design:type", Object)
], FrontendApplication.prototype, "graphProvider", void 0);
__decorate([
    inject(IGraphCommandService),
    __metadata("design:type", Object)
], FrontendApplication.prototype, "commandService", void 0);
__decorate([
    inject(IModelService),
    __metadata("design:type", Object)
], FrontendApplication.prototype, "modelService", void 0);
FrontendApplication = __decorate([
    singleton()
], FrontendApplication);
export { FrontendApplication };
