// @antv/l7plot/dist/lib/core/plot/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Plot = void 0;
var tslib_1 = require("tslib");
var map_1 = require("../map");
var utils_1 = require("../../utils");
var text_layer_1 = require("../../layers/text-layer");
var types_1 = require("../../types");
var source_1 = require("../../adaptor/source");
var util_1 = require("@antv/util");
var DEFAULT_OPTIONS = {
    autoFit: false,
};
var Plot = /** @class */ (function (_super) {
    (0, tslib_1.__extends)(Plot, _super);
    function Plot(container, options) {
        var _this = this;
        if (typeof container === 'string' || container instanceof Element) {
            if (options === undefined) {
                throw new Error('options is undefined');
            }
            _this = _super.call(this, options) || this;
            _this.container = _this.createContainer(container);
            _this.theme = _this.createTheme();
            _this.scene = _this.createScene();
            _this.registerResources();
            _this.initLayers();
        }
        else {
            _this = _super.call(this, container) || this;
        }
        return _this;
    }
    /**
     * 初始化图层
     */
    Plot.prototype.initLayers = function () {
        this.source = this.createSource();
        this.render();
        this.inited = true;
    };
    /**
     * 初始化图层事件
     */
    Plot.prototype.initLayersEvent = function () {
        //
    };
    /**
     * 获取默认配置
     */
    Plot.prototype.getDefaultOptions = function () {
        return Plot.DefaultOptions;
    };
    /**
     * 创建 source 实例
     */
    Plot.prototype.createSource = function () {
        var _a = this.options.source, data = _a.data, aggregation = _a.aggregation, sourceCFG = (0, tslib_1.__rest)(_a, ["data", "aggregation"]);
        aggregation && source_1.MappingSource.aggregation(sourceCFG, aggregation);
        var source = new types_1.Source(data, sourceCFG);
        return source;
    };
    /**
     * 创建数据标签图层
     */
    Plot.prototype.createLabelLayer = function (source, label, plotLayerConfig) {
        var _a = plotLayerConfig || {}, visible = _a.visible, minZoom = _a.minZoom, maxZoom = _a.maxZoom, _b = _a.zIndex, zIndex = _b === void 0 ? 0 : _b;
        var textLayer = new text_layer_1.TextLayer((0, tslib_1.__assign)({ name: 'labelLayer', visible: visible, minZoom: minZoom, maxZoom: maxZoom, zIndex: zIndex + 0.1, source: source }, label));
        return textLayer;
    };
    /**
     * 渲染
     */
    Plot.prototype.render = function () {
        var layerGroup = this.createLayers(this.source);
        if (this.inited) {
            this.layerGroup.removeAllLayer();
            layerGroup.addTo(this.scene);
            this.layerGroup = layerGroup;
            this.updateComponents();
        }
        else {
            this.layerGroup = layerGroup;
            this.onLayersLoaded();
            layerGroup.addTo(this.scene);
        }
        this.initLayersEvent();
    };
    /**
     * 图表图层加载成功
     */
    Plot.prototype.onLayersLoaded = function () {
        var _this = this;
        var onLoaded = function () {
            _this.initComponents();
            _this.loaded = true;
            _this.emit('loaded');
        };
        if (this.scene['sceneService'].loaded) {
            this.sceneLoaded = true;
            this.layersLoaded && onLoaded();
        }
        else {
            this.scene.once('loaded', function () {
                _this.sceneLoaded = true;
                _this.layersLoaded && onLoaded();
            });
        }
        if (this.layerGroup.isEmpty()) {
            this.layersLoaded = true;
        }
        else {
            this.layerGroup.once('inited-all', function () {
                _this.layersLoaded = true;
                _this.sceneLoaded && onLoaded();
            });
        }
    };
    /**
     * 挂载到容器
     */
    Plot.prototype.attachToScene = function (scene, theme) {
        this.scene = scene;
        this.theme = theme;
        this.initLayers();
    };
    /**
     * 更新: 更新配置且重新渲染
     */
    Plot.prototype.update = function (options) {
        this.updateOption(options);
        if (options.map && !(0, util_1.isEqual)(this.lastOptions.map, this.options.map)) {
            this.updateMap(options.map);
        }
        if (options.source && !(0, util_1.isEqual)(this.lastOptions.source, this.options.source)) {
            var _a = options.source, data = _a.data, sourceConfig = (0, tslib_1.__rest)(_a, ["data"]);
            this.changeData(data, sourceConfig);
        }
        this.render();
    };
    /**
     * 更新: 更新数据
     */
    Plot.prototype.changeData = function (data, cfg) {
        this.options.source = (0, utils_1.deepAssign)({}, this.options.source, (0, tslib_1.__assign)({ data: data }, cfg));
        var _a = this.options.source, aggregation = _a.aggregation, sourceCFG = (0, tslib_1.__rest)(_a, ["aggregation"]);
        aggregation && source_1.MappingSource.aggregation(sourceCFG, aggregation);
        this.source.setData(this.options.source.data, sourceCFG);
        // 更新 legend
        if (this.options.legend) {
            this.updateLegendControl(this.options.legend);
        }
    };
    /**
     * 默认的 options 配置项
     */
    Plot.DefaultOptions = DEFAULT_OPTIONS;
    /**
     * 地图图表类型
     */
    Plot.PlotType = types_1.PlotType;
    return Plot;
}(map_1.Map));
exports.Plot = Plot;
