// @antv/xflow-core/es/command-contributions/cmd-context.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { injectable, inject, postConstruct } from 'mana-syringe';
import { DisposableCollection } from '../common/disposable';
import { IGraphProvider } from '../xflow-main/graph/graph-provider';
import { IModelService, MODELS } from '../model-service';
import { IHookService } from '../hooks';
import { IGraphCommandService } from '../command/interface';
let CmdContext = class CmdContext {
    constructor() {
        /** undos 存在这里 */
        this.toDispose = new DisposableCollection();
        /** command handler的runtimeHook */
        this.runtimeHooks = [];
        /** 获取 x6 实例 */
        this.getX6Graph = () => __awaiter(this, void 0, void 0, function* () {
            if (this.graph)
                return this.graph;
            const graphInstance = yield this.graphProvider.getGraphInstance();
            this.graph = graphInstance;
            return graphInstance;
        });
        /** 获取 graph */
        this.getGraphConfig = () => __awaiter(this, void 0, void 0, function* () {
            return this.graphProvider.getGraphOptions();
        });
        this.getGraphMeta = () => __awaiter(this, void 0, void 0, function* () {
            const service = this.getModelService();
            const meta = this.graphMeta || (yield MODELS.GRAPH_META.useValue(service));
            return meta;
        });
        /** 设置参数 */
        this.setArgs = (args, runtimeHooks = []) => {
            this.args = args;
            this.runtimeHooks = runtimeHooks;
            return { args, hooks: runtimeHooks };
        };
        /** 获取参数 */
        this.getArgs = () => {
            /** 注入graph meta */
            const args = Object.assign(Object.assign({}, this.args), { graphMeta: this.graphMeta, modelService: this.getModelService(), commandService: this.getCommands(), getX6Graph: this.getX6Graph, getGraphConfig: this.getGraphConfig });
            return { args: args, hooks: this.runtimeHooks };
        };
        /** 设置执行结果 */
        this.setResult = (result) => {
            this.result = result;
            return this.result;
        };
        /** 获取执行结果 */
        this.getResult = () => {
            return this.result;
        };
        /** hooks */
        this.getHooks = () => {
            return this.hookService.hookProvider();
        };
        /** 获取Command Service */
        this.getCommands = () => {
            return this.commandService;
        };
        /** 获取Context Service */
        this.getModelService = () => {
            return this.modelService;
        };
        /** 添加undo */
        this.addUndo = (disposable) => {
            if (!Array.isArray(disposable)) {
                return this.addUndo([disposable]);
            }
            this.toDispose.pushAll(disposable);
        };
        /** 执行undo */
        this.undo = () => __awaiter(this, void 0, void 0, function* () {
            yield this.toDispose.dispose();
        });
        /** 是否可以执行undo */
        this.isUndoable = () => {
            return !this.toDispose.disposed;
        };
        /** 获取 toDispose */
        this.getDisposables = () => this.toDispose;
        /** 设置的共享变量 可以在command间共享 */
        this.setGlobal = (key, value) => {
            this.commandService.setGlobal(key, value);
        };
        /** 获取共享变量 */
        this.getGlobal = (key) => {
            return this.commandService.getGlobal(key);
        };
    }
    init() {
        this.getGraphMeta().then(meta => {
            this.graphMeta = meta;
        });
    }
};
__decorate([
    inject(IGraphProvider),
    __metadata("design:type", Object)
], CmdContext.prototype, "graphProvider", void 0);
__decorate([
    inject(IHookService),
    __metadata("design:type", Object)
], CmdContext.prototype, "hookService", void 0);
__decorate([
    inject(IGraphCommandService),
    __metadata("design:type", Object)
], CmdContext.prototype, "commandService", void 0);
__decorate([
    inject(IModelService),
    __metadata("design:type", Object)
], CmdContext.prototype, "modelService", void 0);
__decorate([
    postConstruct(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], CmdContext.prototype, "init", null);
CmdContext = __decorate([
    injectable()
], CmdContext);
export { CmdContext };
