// @antv/xflow-core/lib/common/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.insertCss = exports.throttle = exports.debounce = exports.disposableNoop = exports.noop = void 0;
var noop = function () { return null; };
exports.noop = noop;
exports.disposableNoop = {
    dispose: exports.noop,
};
function debounce(cb, time) {
    var flag;
    return function () {
        clearTimeout(flag);
        flag = window.setTimeout(cb, time);
    };
}
exports.debounce = debounce;
function throttle(cb) {
    var flag;
    return function () {
        if (!flag) {
            flag = requestAnimationFrame(function () {
                flag = null;
                cb();
            });
        }
    };
}
exports.throttle = throttle;
var insertCss = function (cssString) {
    var $style = document.createElement('style');
    $style.setAttribute('type', 'text/css');
    document.head.appendChild($style);
    // This is required for IE8 and below.
    // @ts-ignore
    if ($style.styleSheet) {
        // @ts-ignore
        $style.styleSheet.cssText = cssString;
    }
    else {
        $style.appendChild(document.createTextNode(cssString));
    }
    return $style;
};
exports.insertCss = insertCss;
