// @antv/xflow-core/lib/command-contributions/edge/edge-add.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AddEdgeCommand = exports.NsAddEdge = void 0;
var mana_syringe_1 = require("mana-syringe");
var interface_1 = require("../../command/interface");
var constant_1 = require("../constant");
var disposable_1 = require("../../common/disposable");
var NsAddEdge;
(function (NsAddEdge) {
    /** Command: 用于注册named factory */
    NsAddEdge.command = constant_1.XFlowEdgeCommands.ADD_EDGE;
    /** hookName */
    NsAddEdge.hookKey = 'addEdge';
    /** edge id 类型 */
    NsAddEdge.createEdgeId = function (edge) {
        return edge.source + ":" + edge.sourcePortId + "-" + edge.target + ":" + edge.targetPortId;
    };
})(NsAddEdge = exports.NsAddEdge || (exports.NsAddEdge = {}));
var AddEdgeCommand = /** @class */ (function () {
    /** 创建节点命令 */
    function AddEdgeCommand() {
        var _this = this;
        /** 处理edgeConfig的兜底逻辑 */
        this.processEdgeConfig = function (args, edge) { return __awaiter(_this, void 0, void 0, function () {
            var _a, createIdService, _b;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        if (!!edge.id) return [3 /*break*/, 2];
                        _a = args.createIdService, createIdService = _a === void 0 ? NsAddEdge.createEdgeId : _a;
                        _b = edge;
                        return [4 /*yield*/, createIdService(edge)];
                    case 1:
                        _b.id = _c.sent();
                        _c.label = 2;
                    case 2: return [2 /*return*/, edge];
                }
            });
        }); };
        /** 执行Cmd */
        this.execute = function () { return __awaiter(_this, void 0, void 0, function () {
            var _a, args, runtimeHook, hooks, result;
            var _this = this;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this.ctx.getArgs(), args = _a.args, runtimeHook = _a.hooks;
                        hooks = this.ctx.getHooks();
                        return [4 /*yield*/, hooks.addEdge.call(
                            /** 执行 hooks pipeline处理args */
                            args, 
                            /** 执行 callback */
                            function (handlerArgs) { return __awaiter(_this, void 0, void 0, function () {
                                var cellFactory, createEdgeService, commandService, edgeConfig, _a, graph, edgeCell, cell, undo;
                                return __generator(this, function (_b) {
                                    switch (_b.label) {
                                        case 0:
                                            cellFactory = handlerArgs.cellFactory, createEdgeService = handlerArgs.createEdgeService, commandService = handlerArgs.commandService;
                                            if (!createEdgeService) return [3 /*break*/, 2];
                                            return [4 /*yield*/, createEdgeService(handlerArgs)];
                                        case 1:
                                            _a = _b.sent();
                                            return [3 /*break*/, 3];
                                        case 2:
                                            _a = handlerArgs.edgeConfig;
                                            _b.label = 3;
                                        case 3:
                                            edgeConfig = _a;
                                            return [4 /*yield*/, this.ctx.getX6Graph()];
                                        case 4:
                                            graph = _b.sent();
                                            return [4 /*yield*/, this.processEdgeConfig(handlerArgs, edgeConfig)];
                                        case 5:
                                            _b.sent();
                                            if (!cellFactory) return [3 /*break*/, 7];
                                            return [4 /*yield*/, cellFactory(edgeConfig, this)];
                                        case 6:
                                            cell = _b.sent();
                                            edgeCell = graph.addEdge(cell);
                                            return [3 /*break*/, 8];
                                        case 7:
                                            edgeCell = graph.addEdge(__assign(__assign({}, edgeConfig), { 
                                                /** 由于X6的实现是React节点挂在label上的, 所以必须要给label设置值 */
                                                label: (edgeConfig === null || edgeConfig === void 0 ? void 0 : edgeConfig.label) || edgeConfig, data: __assign({}, edgeConfig) }));
                                            _b.label = 8;
                                        case 8:
                                            undo = disposable_1.Disposable.create(function () {
                                                commandService.executeCommand(constant_1.XFlowEdgeCommands.DEL_EDGE.id, {
                                                    x6Edge: edgeCell,
                                                });
                                            });
                                            /** add undo */
                                            this.ctx.addUndo(undo);
                                            return [2 /*return*/, { edgeConfig: edgeConfig, edgeCell: edgeCell }];
                                    }
                                });
                            }); }, runtimeHook)];
                    case 1:
                        result = _b.sent();
                        this.ctx.setResult(result);
                        return [2 /*return*/, this];
                }
            });
        }); };
        /** undo cmd */
        this.undo = function () { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.ctx.undo();
                return [2 /*return*/, this];
            });
        }); };
        /** redo cmd */
        this.redo = function () { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!!this.isUndoable) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.execute()];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2: return [2 /*return*/, this];
                }
            });
        }); };
    }
    AddEdgeCommand.prototype.init = function () {
        this.ctx = this.contextProvider();
    };
    AddEdgeCommand.prototype.isUndoable = function () {
        return this.ctx.isUndoable();
    };
    __decorate([
        (0, mana_syringe_1.inject)(interface_1.ICommandContextProvider),
        __metadata("design:type", Object)
    ], AddEdgeCommand.prototype, "contextProvider", void 0);
    __decorate([
        (0, mana_syringe_1.postConstruct)(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], AddEdgeCommand.prototype, "init", null);
    AddEdgeCommand = __decorate([
        (0, mana_syringe_1.injectable)({
            token: { token: interface_1.ICommandHandler, named: NsAddEdge.command.id },
        })
        /** 创建节点命令 */
    ], AddEdgeCommand);
    return AddEdgeCommand;
}());
exports.AddEdgeCommand = AddEdgeCommand;
