// @antv/xflow-extension/lib/canvas-mini-map/x6-node-view/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimpleNodeView = void 0;
var x6_1 = require("@antv/x6");
var SimpleNodeView = /** @class */ (function (_super) {
    __extends(SimpleNodeView, _super);
    function SimpleNodeView() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SimpleNodeView.prototype.renderMarkup = function () {
        return this.renderJSONMarkup({
            tagName: 'rect',
            selector: 'body',
        });
    };
    SimpleNodeView.prototype.update = function () {
        _super.prototype.update.call(this, {
            body: {
                refWidth: '100%',
                refHeight: '100%',
                fill: SimpleNodeView.nodeFillColor,
            },
        });
    };
    SimpleNodeView.nodeFillColor = '#31d0c6';
    SimpleNodeView.setNodeFillColor = function (color) {
        SimpleNodeView.nodeFillColor = color;
    };
    return SimpleNodeView;
}(x6_1.NodeView));
exports.SimpleNodeView = SimpleNodeView;
