// @antv/xflow-extension/lib/canvas-dag-extension/contributions/command.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.execCmd = exports.QueryGraphStatusCommand = exports.NsGraphStatusCommand = void 0;
var xflow_core_1 = require("@antv/xflow-core");
var constants_1 = require("../constants");
var NsGraphStatusCommand;
(function (NsGraphStatusCommand) {
    var _this = this;
    /** Command: 获取执行状态 */
    NsGraphStatusCommand.MODEL = constants_1.GRAPH_STATUS_INFO;
    /** Command: 用于注册 named factory */
    NsGraphStatusCommand.command = constants_1.XFlowDagCommands.QUERY_GRAPH_STATUS;
    /** hookName */
    NsGraphStatusCommand.hookKey = 'queryGraphStatus';
    /** 状态 类型 */
    var StatusEnum;
    (function (StatusEnum) {
        StatusEnum["SUCCESS"] = "success";
        StatusEnum["PROCESSING"] = "processing";
        StatusEnum["ERROR"] = "error";
        StatusEnum["DEFAULT"] = "default";
        StatusEnum["WARNING"] = "warning";
    })(StatusEnum = NsGraphStatusCommand.StatusEnum || (NsGraphStatusCommand.StatusEnum = {}));
    /** 按status 分类  */
    NsGraphStatusCommand.initStatusMap = function () {
        var _a;
        return (_a = {},
            _a[NsGraphStatusCommand.StatusEnum.DEFAULT] = [],
            _a[NsGraphStatusCommand.StatusEnum.PROCESSING] = [],
            _a[NsGraphStatusCommand.StatusEnum.ERROR] = [],
            _a[NsGraphStatusCommand.StatusEnum.WARNING] = [],
            _a[NsGraphStatusCommand.StatusEnum.SUCCESS] = [],
            _a);
    };
    /** 接口返回 类型 */
    NsGraphStatusCommand.groupByStatus = function (data) {
        var seed = NsGraphStatusCommand.initStatusMap();
        return Object.entries(data).reduce(function (acc, _a) {
            var nodeId = _a[0], value = _a[1];
            acc[value.status].push(nodeId);
            return acc;
        }, seed);
    };
    /** diff status */
    NsGraphStatusCommand.statusDiff = function (cur, next) {
        var items = new Set(__spreadArray(__spreadArray([], cur, true), next, true));
        var diff = { current: [], add: [], remove: [] };
        items.forEach(function (item) {
            if (next.includes(item) && !cur.includes(item)) {
                diff.add.push(item);
            }
            if (!next.includes(item) && cur.includes(item)) {
                diff.remove.push(item);
            }
        });
        return diff;
    };
    NsGraphStatusCommand.shouldStop = function (info) { return __awaiter(_this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, [StatusEnum.ERROR, StatusEnum.SUCCESS].includes(info.graphStatus)];
        });
    }); };
})(NsGraphStatusCommand = exports.NsGraphStatusCommand || (exports.NsGraphStatusCommand = {}));
/** 创建节点命令 */
var QueryGraphStatusCommand = /** @class */ (function () {
    function QueryGraphStatusCommand() {
        var _this = this;
        /** 状态缓存 */
        this.statusInfo = NsGraphStatusCommand.initStatusMap();
        /** 获取Model */
        this.getStatusModel = function () { return __awaiter(_this, void 0, void 0, function () {
            var ctx, modelService, statusModel;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        ctx = this.contextProvider();
                        modelService = ctx.getModelService();
                        return [4 /*yield*/, constants_1.GRAPH_STATUS_INFO.getModel(modelService)];
                    case 1:
                        statusModel = _a.sent();
                        return [2 /*return*/, statusModel];
                }
            });
        }); };
        /** 更新Model数据 */
        this.updateModelValue = function (callback) { return __awaiter(_this, void 0, void 0, function () {
            var statusModel;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getStatusModel()];
                    case 1:
                        statusModel = _a.sent();
                        statusModel.setValue(callback);
                        return [2 /*return*/];
                }
            });
        }); };
        /** 停止Looping的flag */
        this.stopCurrentStatusLooping = function () { return __awaiter(_this, void 0, void 0, function () {
            var statusModel, subscription;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getStatusModel()];
                    case 1:
                        statusModel = _a.sent();
                        return [4 /*yield*/, statusModel.getValidValue()];
                    case 2:
                        subscription = (_a.sent()).subscription;
                        subscription.dispose();
                        return [2 /*return*/];
                }
            });
        }); };
        /** 设置停止looping的方法 */
        this.addStopDispose = function () { return __awaiter(_this, void 0, void 0, function () {
            var statusModel, subscription;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getStatusModel()];
                    case 1:
                        statusModel = _a.sent();
                        return [4 /*yield*/, statusModel.getValidValue()];
                    case 2:
                        subscription = (_a.sent()).subscription;
                        subscription.push({
                            dispose: function () {
                                /** 重制processing的边的状态 */
                                _this.updateEdges(_this.statusInfo, NsGraphStatusCommand.initStatusMap());
                                _this.isLooping = false;
                            },
                        });
                        return [2 /*return*/];
                }
            });
        }); };
        /** 更新Graph的数据 */
        this.updateGraph = function (curStatusInfo, nextStatusMap, statusMap) {
            /** 更新节点 */
            _this.updateNodes(curStatusInfo, nextStatusMap, statusMap);
            /** 更新边 */
            _this.updateEdges(curStatusInfo, nextStatusMap);
        };
        /** 更新节点数据 */
        this.updateNodeData = function (id, data) {
            var cell = _this.x6Graph.getCellById(id);
            if (!cell) {
                return;
            }
            cell.setData(__assign(__assign(__assign(__assign({}, cell.getData()), cell.getSize()), cell.getPosition()), data));
        };
        /** 更新连线数据 */
        this.updateNodes = function (curStatusInfo, nextStatusMap, statusMap) {
            Object.keys(curStatusInfo).forEach(function (status) {
                var nodes = NsGraphStatusCommand.statusDiff(curStatusInfo[status], nextStatusMap[status]);
                nodes.add.forEach(function (id) {
                    _this.updateNodeData(id, statusMap[id]);
                });
                nodes.remove.forEach(function (id) {
                    _this.updateNodeData(id, statusMap[id]);
                });
            });
        };
        /** 更新节点数据 */
        this.updateEdges = function (curStatusInfo, nextStatusMap) {
            var runningNodes = NsGraphStatusCommand.statusDiff(curStatusInfo.processing, nextStatusMap.processing);
            var edges = _this.x6Graph.getEdges();
            edges.forEach(function (edge) {
                var _a;
                var view = (_a = _this.x6Graph) === null || _a === void 0 ? void 0 : _a.findViewByCell(edge.id);
                var target = edge.getTargetCellId();
                if (!target) {
                    return;
                }
                var targetNodeId = target.toString();
                if (!view) {
                    return;
                }
                if (runningNodes.add.includes(targetNodeId)) {
                    /** 新增className */
                    return view.addClass(constants_1.EDGE_PROCESSING_CLASSNAME);
                }
                else if (runningNodes.remove.includes(targetNodeId)) {
                    /** 移除className */
                    return view.removeClass(constants_1.EDGE_PROCESSING_CLASSNAME);
                }
            });
        };
        /** 执行Cmd */
        this.execute = function () { return __awaiter(_this, void 0, void 0, function () {
            var ctx, _a, args, runtimeHook, hooks, _b, result;
            var _this = this;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        ctx = this.contextProvider();
                        _a = ctx.getArgs(), args = _a.args, runtimeHook = _a.hooks;
                        hooks = ctx.getHooks();
                        // 引用
                        _b = this;
                        return [4 /*yield*/, ctx.getX6Graph()];
                    case 1:
                        // 引用
                        _b.x6Graph = _c.sent();
                        return [4 /*yield*/, hooks.queryGraphStatus.call(args, function (handlerArgs) { return __awaiter(_this, void 0, void 0, function () {
                                var graphStatusService, _a, shouldStop, doOnce, loopFunction;
                                var _this = this;
                                return __generator(this, function (_b) {
                                    switch (_b.label) {
                                        case 0:
                                            graphStatusService = handlerArgs.graphStatusService, _a = handlerArgs.shouldStop, shouldStop = _a === void 0 ? NsGraphStatusCommand.shouldStop : _a, doOnce = handlerArgs.doOnce;
                                            // 取消已有的循环
                                            return [4 /*yield*/, this.stopCurrentStatusLooping()
                                                // 循环函数
                                            ];
                                        case 1:
                                            // 取消已有的循环
                                            _b.sent();
                                            loopFunction = function (service, loopInterval) {
                                                if (loopInterval === void 0) { loopInterval = 10000; }
                                                return __awaiter(_this, void 0, void 0, function () {
                                                    var data, statusGroupMap, isStop;
                                                    return __generator(this, function (_a) {
                                                        switch (_a.label) {
                                                            case 0: return [4 /*yield*/, service(handlerArgs)
                                                                // 更新全局状态
                                                            ];
                                                            case 1:
                                                                data = _a.sent();
                                                                // 更新全局状态
                                                                this.updateModelValue(function (state) {
                                                                    Object.entries(data.statusMap).forEach(function (_a) {
                                                                        var key = _a[0], value = _a[1];
                                                                        state.statusMap.set(key, value);
                                                                    });
                                                                    state.graphStatus = data.graphStatus;
                                                                });
                                                                statusGroupMap = NsGraphStatusCommand.groupByStatus(data.statusMap);
                                                                this.updateGraph(this.statusInfo, statusGroupMap, data.statusMap);
                                                                // 缓存当前状态作为对比数据
                                                                this.statusInfo = statusGroupMap;
                                                                // 延迟
                                                                return [4 /*yield*/, (0, xflow_core_1.delay)(loopInterval)
                                                                    // 判断是否循环
                                                                ];
                                                            case 2:
                                                                // 延迟
                                                                _a.sent();
                                                                // 判断是否循环
                                                                if (!this.isLooping) {
                                                                    return [2 /*return*/];
                                                                }
                                                                if (!shouldStop) return [3 /*break*/, 4];
                                                                return [4 /*yield*/, shouldStop(data, handlerArgs)];
                                                            case 3:
                                                                isStop = _a.sent();
                                                                if (isStop) {
                                                                    return [2 /*return*/];
                                                                }
                                                                _a.label = 4;
                                                            case 4: 
                                                            // 执行下次调用
                                                            return [4 /*yield*/, loopFunction(service, loopInterval)];
                                                            case 5:
                                                                // 执行下次调用
                                                                _a.sent();
                                                                return [2 /*return*/];
                                                        }
                                                    });
                                                });
                                            };
                                            // 从参数更新循环的FLAG
                                            this.isLooping = handlerArgs.loop === undefined ? true : handlerArgs.loop;
                                            // 添加可以停止loop的回调
                                            return [4 /*yield*/, this.addStopDispose()
                                                // 执行
                                            ];
                                        case 2:
                                            // 添加可以停止loop的回调
                                            _b.sent();
                                            if (!doOnce) return [3 /*break*/, 4];
                                            return [4 /*yield*/, doOnce(handlerArgs)];
                                        case 3:
                                            _b.sent();
                                            _b.label = 4;
                                        case 4:
                                            loopFunction(graphStatusService, handlerArgs.loopInterval);
                                            return [2 /*return*/, {}];
                                    }
                                });
                            }); }, runtimeHook)];
                    case 2:
                        result = _c.sent();
                        ctx.setResult(result);
                        return [2 /*return*/, this];
                }
            });
        }); };
        /** undo cmd */
        this.undo = function () { return __awaiter(_this, void 0, void 0, function () {
            var ctx;
            return __generator(this, function (_a) {
                ctx = this.contextProvider();
                if (this.isUndoable()) {
                    ctx.undo();
                }
                return [2 /*return*/, this];
            });
        }); };
        /** redo cmd */
        this.redo = function () { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!!this.isUndoable()) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.execute()];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2: return [2 /*return*/, this];
                }
            });
        }); };
    }
    QueryGraphStatusCommand.prototype.isUndoable = function () {
        var ctx = this.contextProvider();
        return ctx.isUndoable();
    };
    __decorate([
        xflow_core_1.ManaSyringe.inject(xflow_core_1.ICommandContextProvider),
        __metadata("design:type", Object)
    ], QueryGraphStatusCommand.prototype, "contextProvider", void 0);
    QueryGraphStatusCommand = __decorate([
        xflow_core_1.ManaSyringe.injectable({
            token: { token: xflow_core_1.ICommandHandler, named: NsGraphStatusCommand.command.id },
        })
    ], QueryGraphStatusCommand);
    return QueryGraphStatusCommand;
}());
exports.QueryGraphStatusCommand = QueryGraphStatusCommand;
var execCmd = function () { };
exports.execCmd = execCmd;
