// @antv/xflow-core/es/command-contributions/group/group-del.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { inject, injectable } from 'mana-syringe';
import { ICommandHandler, ICommandContextProvider } from '../../command/interface';
import { XFlowGroupCommands } from '../constant';
import { Disposable } from '../../common/disposable';
import { XFlowNodeCommands } from '../constant';
export var NsDelGroup;
(function (NsDelGroup) {
    /** Command: 用于注册named factory */
    NsDelGroup.command = XFlowGroupCommands.DEL_GROUP;
    /** hookName */
    NsDelGroup.hookKey = 'delGroup';
})(NsDelGroup || (NsDelGroup = {}));
let DelGroupCommand = 
/** 创建节点命令 */
class DelGroupCommand {
    constructor() {
        /** 执行Cmd */
        this.execute = () => __awaiter(this, void 0, void 0, function* () {
            const ctx = this.contextProvider();
            const { args, hooks: runtimeHook } = ctx.getArgs();
            const hooks = ctx.getHooks();
            const result = yield hooks.delGroup.call(args, (handlerArgs) => __awaiter(this, void 0, void 0, function* () {
                const graph = yield ctx.getX6Graph();
                const { nodeConfig, commandService, deleteService: deleteGroupService } = handlerArgs;
                const { id } = nodeConfig;
                const node = graph.getCellById(id);
                if (deleteGroupService) {
                    const canDel = yield deleteGroupService(handlerArgs);
                    if (!canDel)
                        return { err: 'service rejected' };
                }
                if (!node) {
                    return { err: 'target node is not exist' };
                }
                // 不是Group的节点不能解散
                if (node.getProp('isGroup') !== true) {
                    return { err: 'target node is not group' };
                }
                const { isCollapsed } = node.getData();
                if (isCollapsed) {
                    yield commandService.executeCommand(XFlowGroupCommands.COLLAPSE_GROUP.id, {
                        nodeId: node.id,
                        isCollapsed: false,
                        collapsedSize: { width: 0, height: 0 },
                    });
                }
                const childrens = node.getChildren();
                if (childrens) {
                    childrens.forEach(child => {
                        node.unembed(child);
                        child.prop('group', '');
                        child.setData(Object.assign(Object.assign({}, child.getData()), { group: '' }));
                    });
                }
                yield commandService.executeCommand(XFlowNodeCommands.DEL_NODE.id, {
                    nodeConfig,
                });
                /** add undo: delete node */
                ctx.addUndo(Disposable.create(() => __awaiter(this, void 0, void 0, function* () {
                    commandService.executeCommand(XFlowGroupCommands.ADD_GROUP.id, {
                        nodeConfig,
                    });
                })));
                return { err: null };
            }), runtimeHook);
            ctx.setResult(result);
            return this;
        });
        /** undo cmd */
        this.undo = () => __awaiter(this, void 0, void 0, function* () {
            const ctx = this.contextProvider();
            if (this.isUndoable()) {
                ctx.undo();
            }
            return this;
        });
        /** redo cmd */
        this.redo = () => __awaiter(this, void 0, void 0, function* () {
            if (!this.isUndoable()) {
                yield this.execute();
            }
            return this;
        });
    }
    isUndoable() {
        const ctx = this.contextProvider();
        return ctx.isUndoable();
    }
};
__decorate([
    inject(ICommandContextProvider),
    __metadata("design:type", Object)
], DelGroupCommand.prototype, "contextProvider", void 0);
DelGroupCommand = __decorate([
    injectable({
        token: { token: ICommandHandler, named: NsDelGroup.command.id },
    })
    /** 创建节点命令 */
], DelGroupCommand);
export { DelGroupCommand };
