// @antv/xflow-extension/lib/flowchart-node-panel/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FlowchartNodePanel = exports.NodePanelMain = exports.setGroupRender = void 0;
var react_1 = __importStar(require("react"));
var xflow_core_1 = require("@antv/xflow-core");
var icons_1 = require("@ant-design/icons");
var base_panel_1 = require("../base-panel");
var panel_body_1 = require("./panel-body");
// import { NodePanelHeader } from '../canvas-node-tree-panel/panel-header'
var panel_header_1 = require("./panel-header");
var utils_1 = require("../canvas-node-tree-panel/utils");
var service_1 = require("./service");
var constants_1 = require("./constants");
var group_panel_1 = require("./group-panel");
Object.defineProperty(exports, "setGroupRender", { enumerable: true, get: function () { return group_panel_1.setGroupRender; } });
__exportStar(require("./constants"), exports);
__exportStar(require("./utils"), exports);
__exportStar(require("./interface"), exports);
require("./style/index");
var NodePanelMain = function (props) {
    var prefixClz = props.prefixClz, _a = props.position, position = _a === void 0 ? { width: 240, top: 0, bottom: 0, left: 0 } : _a, _b = props.showHeader, showHeader = _b === void 0 ? true : _b, rest = __rest(props, ["prefixClz", "position", "showHeader"]);
    var _c = position.width, width = _c === void 0 ? 200 : _c;
    var _d = (0, utils_1.usePanelLyaoutStyle)(props), headerStyle = _d.headerStyle, bodyStyle = _d.bodyStyle;
    var _e = (0, service_1.usePanelData)(props), state = _e.state, onKeywordChange = _e.onKeywordChange;
    return (react_1.default.createElement(react_1.default.Fragment, null,
        showHeader && (react_1.default.createElement(base_panel_1.WorkspacePanel, __assign({}, rest, { position: {
                top: 0,
                left: 0,
                height: constants_1.PANEL_HEADER_HEIGHT,
                width: width,
            } }),
            react_1.default.createElement(panel_header_1.NodePanelHeader, __assign({}, props, { state: state, style: headerStyle, onKeywordChange: onKeywordChange })))),
        react_1.default.createElement(base_panel_1.WorkspacePanel, __assign({ className: constants_1.CONTAINER_CLASS + "-nodes" }, rest, { position: __assign(__assign({}, position), { top: showHeader ? constants_1.PANEL_HEADER_HEIGHT : 0 }) }),
            react_1.default.createElement(panel_body_1.NodePanelBody, __assign({}, props, { state: state, style: bodyStyle })))));
};
exports.NodePanelMain = NodePanelMain;
var FlowchartNodePanel = function (props) {
    var prefixClz = (0, xflow_core_1.useXflowPrefixCls)('node-panel');
    var _a = (0, react_1.useState)(false), collpased = _a[0], setCollpased = _a[1];
    var _b = props.show, show = _b === void 0 ? true : _b, _c = props.position, position = _c === void 0 ? { width: 240, top: 40, bottom: 0, left: 0 } : _c, rest = __rest(props, ["show", "position"]);
    if (!show) {
        return null;
    }
    var _d = position.width, width = _d === void 0 ? 200 : _d, left = position.left;
    return (react_1.default.createElement(base_panel_1.WorkspacePanel, { className: constants_1.CONTAINER_CLASS, position: __assign(__assign({}, position), { left: !collpased ? left : -width }) },
        react_1.default.createElement("div", { className: constants_1.CONTAINER_CLASS + "-wrapper" },
            react_1.default.createElement(base_panel_1.WorkspacePanel, __assign({ className: prefixClz }, rest, { position: {
                    top: 0,
                    bottom: 0,
                    left: 0,
                    right: 0,
                } }),
                react_1.default.createElement(exports.NodePanelMain, __assign({}, props, { prefixClz: prefixClz, position: position }))),
            react_1.default.createElement("div", { className: constants_1.CONTAINER_CLASS + "-icon", style: {
                    top: 21,
                    right: !collpased ? -10 : -20,
                    borderRadius: !collpased ? '50%' : '0 50% 50% 0',
                    borderLeft: !collpased ? '' : 'none',
                }, onClick: function () {
                    setCollpased(!collpased);
                } }, collpased ? react_1.default.createElement(icons_1.DoubleRightOutlined, null) : react_1.default.createElement(icons_1.DoubleLeftOutlined, null)))));
};
exports.FlowchartNodePanel = FlowchartNodePanel;
