// @antv/xflow-extension/lib/flowchart-node-panel/panel-body.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodePanelBody = void 0;
var react_1 = __importStar(require("react"));
var antd_1 = require("antd");
var lodash_1 = require("lodash");
var x6_1 = require("@antv/x6");
var xflow_core_1 = require("@antv/xflow-core");
var utils_1 = require("./utils");
var utils_2 = require("../flowchart-canvas/utils");
var panel_body_1 = require("../canvas-node-tree-panel/panel-body");
var Panel = antd_1.Collapse.Panel;
var NodePanelBody = function (props) {
    var x6NodeFactory = props.x6NodeFactory, dndOptions = props.dndOptions, state = props.state, prefixClz = props.prefixClz, registerNode = props.registerNode, _a = props.defaultActiveKey, defaultActiveKey = _a === void 0 ? ['official', 'custom'] : _a;
    var _b = (registerNode !== null && registerNode !== void 0 ? registerNode : {}).title, title = _b === void 0 ? '复制节点' : _b;
    var _c = (0, xflow_core_1.useXFlowApp)(), graphProvider = _c.graphProvider, modelService = _c.modelService, commandService = _c.commandService;
    var _d = react_1.default.useState(), dnd = _d[0], setDnd = _d[1];
    /** 获取graph实例 */
    var _e = react_1.default.useState(), graph = _e[0], setGraph = _e[1];
    graphProvider.getGraphInstance().then(function (x6Graph) {
        setGraph(x6Graph);
    });
    var graphConfig = undefined;
    graphProvider.getGraphOptions().then(function (x6GraphConfig) {
        /** x6GraphConfig 上缺失 setNodeRender 方法  */
        (0, utils_1.setNodeRender)((0, utils_2.getProps)('graphConfig'), (0, lodash_1.get)(props, 'registerNode.nodes', []));
        graphConfig = x6GraphConfig;
    });
    var onNodeDrop = (0, react_1.useCallback)(function (node) { return __awaiter(void 0, void 0, void 0, function () {
        var ports, nodeConfig, args, onAddNode;
        var _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    ports = node.ports;
                    nodeConfig = __assign(__assign({}, node), { id: "node-" + (0, xflow_core_1.uuidv4)(), zIndex: 10, ports: __assign(__assign({}, ports), { items: (_a = ports.items) === null || _a === void 0 ? void 0 : _a.map(function (item) { return (__assign(__assign({}, item), { id: (0, xflow_core_1.uuidv4)() })); }) }) });
                    args = {
                        nodeConfig: nodeConfig,
                    };
                    return [4 /*yield*/, commandService.executeCommand(xflow_core_1.XFlowNodeCommands.ADD_NODE.id, args)];
                case 1:
                    _b.sent();
                    onAddNode = (0, utils_2.getProps)('onAddNode');
                    if (typeof onAddNode === 'function') {
                        onAddNode(nodeConfig);
                    }
                    return [2 /*return*/];
            }
        });
    }); }, [commandService]);
    react_1.default.useEffect(function () {
        if (!graph) {
            return;
        }
        var dndInstance = new x6_1.Addon.Dnd(__assign(__assign({ scaled: false, animation: false }, dndOptions), { target: graph, 
            /** 这里考虑到需要新增群组的需求，不使用x6的getDropNod方法
             * 在validateNode时调用command添加
             */
            validateNode: function (droppingNode) { return __awaiter(void 0, void 0, void 0, function () {
                var nodeConfig;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            nodeConfig = __assign(__assign({}, droppingNode.getData()), droppingNode.getPosition());
                            return [4 /*yield*/, onNodeDrop(nodeConfig)];
                        case 1:
                            _a.sent();
                            return [2 /*return*/, false];
                    }
                });
            }); } }));
        setDnd(dndInstance);
    }, [commandService, dndOptions, graph, modelService, onNodeDrop]);
    var onMouseDown = react_1.default.useCallback(function (nodeConfig) { return function (e) {
        if (!graph || !dnd || !graphConfig) {
            return;
        }
        var renderKey = graphConfig.nodeTypeParser(nodeConfig);
        var _a = nodeConfig.width, width = _a === void 0 ? 180 : _a, _b = nodeConfig.height, height = _b === void 0 ? 40 : _b;
        var reactComponent = graphConfig.nodeRender.get(renderKey);
        var wrappedComponent = (0, xflow_core_1.getNodeReactComponent)(reactComponent, commandService, modelService);
        var nodeData = {
            data: nodeConfig,
            width: width,
            height: height,
            view: graphConfig.graphId,
            component: wrappedComponent,
        };
        var x6Node = x6NodeFactory ? x6NodeFactory(nodeData) : (0, panel_body_1.defaultNodeFactory)(nodeData);
        dnd.start(x6Node, e.nativeEvent);
    }; }, [commandService, dnd, graph, graphConfig, modelService, x6NodeFactory]);
    var renderTree = react_1.default.useCallback(function (list) {
        if (list === void 0) { list = []; }
        return list.map(function (item) {
            var popoverContent = item.popoverContent;
            return (react_1.default.createElement(panel_body_1.NodeTitle, { item: item, key: item.id, onMouseDown: onMouseDown(item), popoverContent: popoverContent, prefixClz: prefixClz, modelService: modelService, commandService: commandService, graphConfig: graphConfig }));
        });
    }, [commandService, graphConfig, modelService, onMouseDown, prefixClz]);
    var customNode = state.nodeList.filter(function (item) { return item.isCustom; });
    var officialNode = state.nodeList.filter(function (item) { return !item.isCustom; });
    var searchCustomNode = state.searchList.filter(function (item) { return item.isCustom; });
    var searchOfficialNode = state.searchList.filter(function (item) { return !item.isCustom; });
    var hasCustomNode = customNode.length > 0;
    return (react_1.default.createElement(react_1.default.Fragment, null,
        react_1.default.createElement("div", { className: prefixClz + "-body" },
            react_1.default.createElement(antd_1.Collapse, { defaultActiveKey: defaultActiveKey, style: { border: 'none' } },
                react_1.default.createElement(Panel, { header: "\u901A\u7528\u8282\u70B9", key: "official", style: { border: 'none' } },
                    !state.keyword && (react_1.default.createElement("div", { className: prefixClz + "-official" }, renderTree(officialNode))),
                    state.searchList.length > 0 && (react_1.default.createElement("div", { className: prefixClz + "-official" }, renderTree(searchOfficialNode)))),
                hasCustomNode && (react_1.default.createElement(Panel, { header: title, key: "custom", style: { border: 'none' } },
                    !state.keyword && (react_1.default.createElement("div", { className: prefixClz + "-custom" }, renderTree(customNode))),
                    state.searchList.length > 0 && (react_1.default.createElement("div", { className: prefixClz + "-custom" }, renderTree(searchCustomNode)))))),
            state.keyword && state.searchList.length === 0 && react_1.default.createElement(antd_1.Empty, { style: { marginTop: '48px' } }))));
};
exports.NodePanelBody = NodePanelBody;
