// @antv/xflow-core/lib/command-contributions/group/group-del.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DelGroupCommand = exports.NsDelGroup = void 0;
var mana_syringe_1 = require("mana-syringe");
var interface_1 = require("../../command/interface");
var constant_1 = require("../constant");
var disposable_1 = require("../../common/disposable");
var constant_2 = require("../constant");
var NsDelGroup;
(function (NsDelGroup) {
    /** Command: 用于注册named factory */
    NsDelGroup.command = constant_1.XFlowGroupCommands.DEL_GROUP;
    /** hookName */
    NsDelGroup.hookKey = 'delGroup';
})(NsDelGroup = exports.NsDelGroup || (exports.NsDelGroup = {}));
var DelGroupCommand = /** @class */ (function () {
    /** 创建节点命令 */
    function DelGroupCommand() {
        var _this = this;
        /** 执行Cmd */
        this.execute = function () { return __awaiter(_this, void 0, void 0, function () {
            var ctx, _a, args, runtimeHook, hooks, result;
            var _this = this;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        ctx = this.contextProvider();
                        _a = ctx.getArgs(), args = _a.args, runtimeHook = _a.hooks;
                        hooks = ctx.getHooks();
                        return [4 /*yield*/, hooks.delGroup.call(args, function (handlerArgs) { return __awaiter(_this, void 0, void 0, function () {
                                var graph, nodeConfig, commandService, deleteGroupService, id, node, canDel, isCollapsed, childrens;
                                var _this = this;
                                return __generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0: return [4 /*yield*/, ctx.getX6Graph()];
                                        case 1:
                                            graph = _a.sent();
                                            nodeConfig = handlerArgs.nodeConfig, commandService = handlerArgs.commandService, deleteGroupService = handlerArgs.deleteService;
                                            id = nodeConfig.id;
                                            node = graph.getCellById(id);
                                            if (!deleteGroupService) return [3 /*break*/, 3];
                                            return [4 /*yield*/, deleteGroupService(handlerArgs)];
                                        case 2:
                                            canDel = _a.sent();
                                            if (!canDel)
                                                return [2 /*return*/, { err: 'service rejected' }];
                                            _a.label = 3;
                                        case 3:
                                            if (!node) {
                                                return [2 /*return*/, { err: 'target node is not exist' }];
                                            }
                                            // 不是Group的节点不能解散
                                            if (node.getProp('isGroup') !== true) {
                                                return [2 /*return*/, { err: 'target node is not group' }];
                                            }
                                            isCollapsed = node.getData().isCollapsed;
                                            if (!isCollapsed) return [3 /*break*/, 5];
                                            return [4 /*yield*/, commandService.executeCommand(constant_1.XFlowGroupCommands.COLLAPSE_GROUP.id, {
                                                    nodeId: node.id,
                                                    isCollapsed: false,
                                                    collapsedSize: { width: 0, height: 0 },
                                                })];
                                        case 4:
                                            _a.sent();
                                            _a.label = 5;
                                        case 5:
                                            childrens = node.getChildren();
                                            if (childrens) {
                                                childrens.forEach(function (child) {
                                                    node.unembed(child);
                                                    child.prop('group', '');
                                                    child.setData(__assign(__assign({}, child.getData()), { group: '' }));
                                                });
                                            }
                                            return [4 /*yield*/, commandService.executeCommand(constant_2.XFlowNodeCommands.DEL_NODE.id, {
                                                    nodeConfig: nodeConfig,
                                                })
                                                /** add undo: delete node */
                                            ];
                                        case 6:
                                            _a.sent();
                                            /** add undo: delete node */
                                            ctx.addUndo(disposable_1.Disposable.create(function () { return __awaiter(_this, void 0, void 0, function () {
                                                return __generator(this, function (_a) {
                                                    commandService.executeCommand(constant_1.XFlowGroupCommands.ADD_GROUP.id, {
                                                        nodeConfig: nodeConfig,
                                                    });
                                                    return [2 /*return*/];
                                                });
                                            }); }));
                                            return [2 /*return*/, { err: null }];
                                    }
                                });
                            }); }, runtimeHook)];
                    case 1:
                        result = _b.sent();
                        ctx.setResult(result);
                        return [2 /*return*/, this];
                }
            });
        }); };
        /** undo cmd */
        this.undo = function () { return __awaiter(_this, void 0, void 0, function () {
            var ctx;
            return __generator(this, function (_a) {
                ctx = this.contextProvider();
                if (this.isUndoable()) {
                    ctx.undo();
                }
                return [2 /*return*/, this];
            });
        }); };
        /** redo cmd */
        this.redo = function () { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!!this.isUndoable()) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.execute()];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2: return [2 /*return*/, this];
                }
            });
        }); };
    }
    DelGroupCommand.prototype.isUndoable = function () {
        var ctx = this.contextProvider();
        return ctx.isUndoable();
    };
    __decorate([
        (0, mana_syringe_1.inject)(interface_1.ICommandContextProvider),
        __metadata("design:type", Object)
    ], DelGroupCommand.prototype, "contextProvider", void 0);
    DelGroupCommand = __decorate([
        (0, mana_syringe_1.injectable)({
            token: { token: interface_1.ICommandHandler, named: NsDelGroup.command.id },
        })
        /** 创建节点命令 */
    ], DelGroupCommand);
    return DelGroupCommand;
}());
exports.DelGroupCommand = DelGroupCommand;
