// @antv/xflow-hook/lib/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HookHub = exports.ScheduleTypeEnum = void 0;
var interface_1 = require("./interface");
var utils_1 = require("./utils");
var interface_2 = require("./interface");
Object.defineProperty(exports, "ScheduleTypeEnum", { enumerable: true, get: function () { return interface_2.ScheduleTypeEnum; } });
var HookHub = /** @class */ (function () {
    function HookHub(options) {
        var _a;
        var _this = this;
        /** scheduleType */
        this.scheduleType = interface_1.ScheduleTypeEnum.ASYNC_SRRIES;
        /** hasRegistered */
        this.hasHook = function (hookName) {
            return _this.hookMap.has(hookName);
        };
        /** getHooks */
        this.getHooks = function (runtimeHooks, sort) {
            if (runtimeHooks === void 0) { runtimeHooks = []; }
            if (sort === void 0) { sort = true; }
            var hooks = utils_1.HookUtils.normalize(runtimeHooks, _this.hookMap);
            if (!sort) {
                return hooks;
            }
            utils_1.HookUtils.sort(hooks, _this.hookMap);
            return hooks;
        };
        /** registerHook */
        this.registerHook = function (hookMeta) {
            if (_this.hookMap.has(hookMeta.name)) {
                console.error(hookMeta.name + " is duplicated in hookmap");
            }
            _this.hookMap.set(hookMeta.name, hookMeta);
            return {
                dispose: function () {
                    _this.hookMap.delete(hookMeta.name);
                },
            };
        };
        /** registerHook */
        this.call = function (args, main, runtimeHook) {
            if (main === void 0) { main = function (mainArgs) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                return [2 /*return*/, mainArgs];
            }); }); }; }
            if (runtimeHook === void 0) { runtimeHook = []; }
            return __awaiter(_this, void 0, void 0, function () {
                var hooks, scheduler;
                return __generator(this, function (_a) {
                    hooks = this.getHooks(runtimeHook);
                    scheduler = this.schedulers[this.scheduleType];
                    return [2 /*return*/, scheduler(args, main, hooks)];
                });
            });
        };
        /** 执行hook的scheduler */
        this.schedulers = (_a = {},
            /** pipeline执行 */
            _a[interface_1.ScheduleTypeEnum.ASYNC_SRRIES] = function (args, main, hooks) {
                if (hooks === void 0) { hooks = []; }
                return __awaiter(_this, void 0, void 0, function () {
                    var callback, _i, hooks_1, hook, result, err;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                callback = main;
                                _i = 0, hooks_1 = hooks;
                                _a.label = 1;
                            case 1:
                                if (!(_i < hooks_1.length)) return [3 /*break*/, 7];
                                hook = hooks_1[_i];
                                if (![0, 1].includes(hook.handler.length)) return [3 /*break*/, 3];
                                return [4 /*yield*/, hook.handler.call(this, args)];
                            case 2:
                                _a.sent();
                                return [3 /*break*/, 6];
                            case 3:
                                if (!([2].includes(hook.handler.length) && callback !== null)) return [3 /*break*/, 5];
                                return [4 /*yield*/, hook.handler.call(this, args, callback)
                                    /** 如果返回为null，则直接中断执行 */
                                ];
                            case 4:
                                result = _a.sent();
                                /** 如果返回为null，则直接中断执行 */
                                if (result === null) {
                                    callback = null;
                                    return [3 /*break*/, 7];
                                }
                                else if (typeof result === 'function') {
                                    callback = result;
                                    return [3 /*break*/, 6];
                                }
                                _a.label = 5;
                            case 5:
                                err = utils_1.ErrorUtils.InvalidHookArguments(hook);
                                throw err;
                            case 6:
                                _i++;
                                return [3 /*break*/, 1];
                            case 7:
                                if (!callback) return [3 /*break*/, 9];
                                return [4 /*yield*/, callback.call(this, args)];
                            case 8: return [2 /*return*/, _a.sent()];
                            case 9: return [2 /*return*/];
                        }
                    });
                });
            },
            _a[interface_1.ScheduleTypeEnum.ASYNC_PARALLEL] = function (args, main, hooks) {
                if (hooks === void 0) { hooks = []; }
                return __awaiter(_this, void 0, void 0, function () {
                    var promises, defer;
                    var _this = this;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                promises = hooks.map(function (hook) {
                                    if ([0, 1].includes(hook.handler.length)) {
                                        return hook.handler.call(_this, args);
                                    }
                                    if ([2].includes(hook.handler.length)) {
                                        return hook.handler.call(_this, args, main);
                                    }
                                    throw utils_1.ErrorUtils.InvalidHookArguments(hook);
                                });
                                defer = new utils_1.Deferred();
                                Promise.all(promises).then(function (res) { return defer.resolve(res); });
                                if (!main) return [3 /*break*/, 2];
                                return [4 /*yield*/, main.call(this, defer)];
                            case 1: return [2 /*return*/, _a.sent()];
                            case 2: return [2 /*return*/];
                        }
                    });
                });
            },
            _a);
        this.hookMap = new Map();
        if (options && options.type) {
            this.scheduleType = options.type;
        }
    }
    return HookHub;
}());
exports.HookHub = HookHub;
