// @antv/xflow-hook/lib/utils.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Deferred = exports.HookUtils = exports.ErrorUtils = void 0;
var toposort_1 = __importDefault(require("toposort"));
var ErrorUtils;
(function (ErrorUtils) {
    var ErrorEnum;
    (function (ErrorEnum) {
        ErrorEnum["INVALID_HOOK_BEFORE"] = "INVALID_HOOK_BEFORE";
        ErrorEnum["INVALID_HOOK_AFTER"] = "INVALID_HOOK_AFTER";
        ErrorEnum["INVALID_HOOK_ARGS_LENGTH"] = "INVALID_HOOK_ARGS_LENGTH";
    })(ErrorEnum = ErrorUtils.ErrorEnum || (ErrorUtils.ErrorEnum = {}));
    var HookError = /** @class */ (function (_super) {
        __extends(HookError, _super);
        function HookError(code, message) {
            var contexts = [];
            for (var _i = 2; _i < arguments.length; _i++) {
                contexts[_i - 2] = arguments[_i];
            }
            var _this = this;
            if (Array.isArray(message)) {
                message = message
                    .filter(function (line) {
                    return !!line;
                })
                    .join(' ');
            }
            _this = _super.call(this, code + ": " + message) || this;
            if (Error.captureStackTrace) {
                ;
                Error.captureStackTrace(_this, HookError);
            }
            _this.code = code;
            _this.contexts = {};
            contexts.forEach(function (context) {
                if (typeof context === 'object') {
                    Object.entries(context).forEach(function (pair) {
                        var key = pair[0], value = pair[1];
                        if (key !== 'key' && value) {
                            _this.contexts[key] = JSON.parse(JSON.stringify(value));
                        }
                    });
                }
            });
            return _this;
        }
        HookError.prototype.toString = function () {
            return this.code + ": " + this.message;
        };
        return HookError;
    }(Error));
    ErrorUtils.HookError = HookError;
    var createError = function (code, message) {
        var args = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            args[_i - 2] = arguments[_i];
        }
        return new (HookError.bind.apply(HookError, __spreadArray([void 0, code, message], args, false)))();
    };
    ErrorUtils.HookBeforeError = function () {
        return createError(ErrorEnum.INVALID_HOOK_BEFORE, [
            'hook require a before dependency but not exist in hookmap',
        ]);
    };
    ErrorUtils.HookAfterError = function () {
        return createError(ErrorEnum.INVALID_HOOK_AFTER, [
            'hook require a after dependency but not exist in hookmap',
        ]);
    };
    ErrorUtils.InvalidHookArguments = function (hook) {
        return createError(ErrorEnum.INVALID_HOOK_ARGS_LENGTH, [
            'hook handlers must have 0 to 2 arguments',
            "but got " + hook.handler.length,
        ]);
    };
})(ErrorUtils = exports.ErrorUtils || (exports.ErrorUtils = {}));
var HookUtils;
(function (HookUtils) {
    /** 处理 runtime hooks */
    HookUtils.normalize = function (hookConfig, hookMap) {
        if (hookConfig === void 0) { hookConfig = []; }
        var runtimeHook = Array.isArray(hookConfig) ? hookConfig : [hookConfig];
        var innerHooks = [];
        hookMap.forEach(function (val) {
            innerHooks.push(val);
        });
        return __spreadArray(__spreadArray([], innerHooks, true), runtimeHook, true).filter(function (item) { return item && !!item.handler; });
    };
    /** 排序 hooks */
    HookUtils.sort = function (hooks, hookMap) {
        if (hooks === void 0) { hooks = []; }
        var edges = [];
        hooks.forEach(function (hook) {
            if (hook.before) {
                if (!hookMap.has(hook.before)) {
                    throw ErrorUtils.HookBeforeError();
                }
                var edge = [hookMap.get(hook.before), hook];
                edges.push(edge);
            }
            if (hook.after) {
                if (!hookMap.has(hook.after)) {
                    throw ErrorUtils.HookBeforeError();
                }
                var edge = [hook, hookMap.get(hook.after)];
                edges.push(edge);
            }
        });
        var sortedHooks = (0, toposort_1.default)(edges);
        sortedHooks.reverse();
        return sortedHooks;
    };
})(HookUtils = exports.HookUtils || (exports.HookUtils = {}));
var Deferred = /** @class */ (function () {
    function Deferred() {
        var _this = this;
        this.isResolved = false;
        this.isRejected = false;
        this.promise = new Promise(function (resolve, reject) {
            _this.resolve = function (args) {
                _this.isResolved = true;
                resolve(args);
            };
            _this.reject = function (args) {
                _this.isRejected = true;
                reject(args);
            };
        });
    }
    return Deferred;
}());
exports.Deferred = Deferred;
