// @antv/xflow-core/lib/command-contributions/graph/graph-render.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GraphRenderCommand = exports.NsGraphRender = exports.NsGraphUtils = void 0;
var lodash_1 = require("lodash");
var mana_syringe_1 = require("mana-syringe");
var constant_1 = require("../constant");
var interface_1 = require("../../command/interface");
var NsGraphUtils;
(function (NsGraphUtils) {
    function isNodeEqual(curNodeConfig, nextNodeConfig) {
        /** XFlow默认的判断节点是否相等的逻辑 */
        return (0, lodash_1.isEqual)(curNodeConfig, nextNodeConfig);
    }
    NsGraphUtils.isNodeEqual = isNodeEqual;
    function isEdgeEqual(curEdgeConfig, nextEdgeConfig) {
        /** XFlow默认的判断边是否相等的逻辑 */
        return (0, lodash_1.isEqual)(curEdgeConfig, nextEdgeConfig);
    }
    NsGraphUtils.isEdgeEqual = isEdgeEqual;
})(NsGraphUtils = exports.NsGraphUtils || (exports.NsGraphUtils = {}));
var NsGraphRender;
(function (NsGraphRender) {
    /** Command: 用于注册named factory */
    NsGraphRender.command = constant_1.XFlowGraphCommands.GRAPH_RENDER;
    /** hookName */
    NsGraphRender.hookKey = 'graphRender';
})(NsGraphRender = exports.NsGraphRender || (exports.NsGraphRender = {}));
var GraphRenderCommand = /** @class */ (function () {
    /** 画布渲染命令 */
    function GraphRenderCommand() {
        var _this = this;
        this.execute = function () { return __awaiter(_this, void 0, void 0, function () {
            var _a, args, runtimeHook, hooks, result;
            var _this = this;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this.ctx.getArgs(), args = _a.args, runtimeHook = _a.hooks;
                        hooks = this.ctx.getHooks();
                        return [4 /*yield*/, hooks.graphRender.call(args, function (handlerArgs) { return __awaiter(_this, void 0, void 0, function () {
                                var x6Graph, graphMeta, beforeRender, graphData, isNodeEqual, isEdgeEqual, afterRender;
                                return __generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0: return [4 /*yield*/, this.ctx.getX6Graph()];
                                        case 1:
                                            x6Graph = _a.sent();
                                            graphMeta = handlerArgs.graphMeta, beforeRender = handlerArgs.beforeRender, graphData = handlerArgs.graphData, isNodeEqual = handlerArgs.isNodeEqual, isEdgeEqual = handlerArgs.isEdgeEqual, afterRender = handlerArgs.afterRender;
                                            /** 如果用户自定义beforeRender方法 */
                                            beforeRender && beforeRender(graphMeta);
                                            return [4 /*yield*/, this.doLoadGraph(x6Graph, graphData, isNodeEqual, isEdgeEqual)
                                                /** 如果用户自定义afterRender方法 */
                                            ];
                                        case 2:
                                            _a.sent();
                                            /** 如果用户自定义afterRender方法 */
                                            afterRender && afterRender(graphData, graphMeta);
                                            return [2 /*return*/, {}];
                                    }
                                });
                            }); }, runtimeHook)
                            /** 设置结果 */
                        ];
                    case 1:
                        result = _b.sent();
                        /** 设置结果 */
                        this.ctx.setResult(result);
                        return [2 /*return*/, this];
                }
            });
        }); };
        this.undo = function () { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.ctx.undo();
                return [2 /*return*/, this];
            });
        }); };
        this.redo = function () { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!!this.ctx.isUndoable) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.execute()];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2: return [2 /*return*/, this];
                }
            });
        }); };
        this.doLoadGraph = function (x6Graph, graphData, isNodeEqual, isEdgeEqual) { return __awaiter(_this, void 0, void 0, function () {
            var commandService, _a, addNodeConfigs, addEdgeConfigs, 
            // retainNodes,
            // retainEdges,
            removeNodes, removeEdges, updateNodes, updateEdges, _i, addNodeConfigs_1, nodeConfig, _b, addEdgeConfigs_1, edgeConfig, _c, removeNodes_1, removeNode, nodeData, _d, removeEdges_1, removeEdge, edgeData, _e, updateNodes_1, updateNode, nodeData, _f, updateEdges_1, updateEdge, edgeData;
            var _this = this;
            return __generator(this, function (_g) {
                switch (_g.label) {
                    case 0:
                        if (!(x6Graph === null || x6Graph === void 0 ? void 0 : x6Graph.isFrozen())) {
                            x6Graph === null || x6Graph === void 0 ? void 0 : x6Graph.freeze();
                        }
                        commandService = this.ctx.getCommands();
                        _a = this.graphDataDiff(x6Graph, graphData, isNodeEqual, isEdgeEqual), addNodeConfigs = _a.addNodeConfigs, addEdgeConfigs = _a.addEdgeConfigs, removeNodes = _a.removeNodes, removeEdges = _a.removeEdges, updateNodes = _a.updateNodes, updateEdges = _a.updateEdges;
                        _i = 0, addNodeConfigs_1 = addNodeConfigs;
                        _g.label = 1;
                    case 1:
                        if (!(_i < addNodeConfigs_1.length)) return [3 /*break*/, 4];
                        nodeConfig = addNodeConfigs_1[_i];
                        return [4 /*yield*/, commandService.executeCommand(constant_1.XFlowNodeCommands.ADD_NODE.id, {
                                nodeConfig: nodeConfig,
                            }, {
                                name: 'remove servcie ',
                                handler: function (args) { return __awaiter(_this, void 0, void 0, function () {
                                    return __generator(this, function (_a) {
                                        delete args.createNodeService;
                                        return [2 /*return*/];
                                    });
                                }); },
                            })];
                    case 2:
                        _g.sent();
                        _g.label = 3;
                    case 3:
                        _i++;
                        return [3 /*break*/, 1];
                    case 4: return [4 /*yield*/, commandService.executeCommand(constant_1.XFlowGroupCommands.INIT_GROUP.id, {
                            graphData: { nodes: addNodeConfigs, edges: [] },
                        })];
                    case 5:
                        _g.sent();
                        _b = 0, addEdgeConfigs_1 = addEdgeConfigs;
                        _g.label = 6;
                    case 6:
                        if (!(_b < addEdgeConfigs_1.length)) return [3 /*break*/, 9];
                        edgeConfig = addEdgeConfigs_1[_b];
                        return [4 /*yield*/, commandService.executeCommand(constant_1.XFlowEdgeCommands.ADD_EDGE.id, {
                                edgeConfig: edgeConfig,
                            }, {
                                name: 'remove servcie ',
                                handler: function (args) { return __awaiter(_this, void 0, void 0, function () {
                                    return __generator(this, function (_a) {
                                        delete args.createEdgeService;
                                        return [2 /*return*/];
                                    });
                                }); },
                            })];
                    case 7:
                        _g.sent();
                        _g.label = 8;
                    case 8:
                        _b++;
                        return [3 /*break*/, 6];
                    case 9:
                        _c = 0, removeNodes_1 = removeNodes;
                        _g.label = 10;
                    case 10:
                        if (!(_c < removeNodes_1.length)) return [3 /*break*/, 15];
                        removeNode = removeNodes_1[_c];
                        nodeData = removeNode === null || removeNode === void 0 ? void 0 : removeNode.getData();
                        if (!nodeData.isGroup) return [3 /*break*/, 12];
                        return [4 /*yield*/, commandService.executeCommand(constant_1.XFlowGroupCommands.DEL_GROUP.id, {
                                nodeConfig: nodeData,
                            })];
                    case 11:
                        _g.sent();
                        return [3 /*break*/, 14];
                    case 12: return [4 /*yield*/, commandService.executeCommand(constant_1.XFlowNodeCommands.DEL_NODE.id, { nodeConfig: nodeData })];
                    case 13:
                        _g.sent();
                        _g.label = 14;
                    case 14:
                        _c++;
                        return [3 /*break*/, 10];
                    case 15:
                        _d = 0, removeEdges_1 = removeEdges;
                        _g.label = 16;
                    case 16:
                        if (!(_d < removeEdges_1.length)) return [3 /*break*/, 19];
                        removeEdge = removeEdges_1[_d];
                        edgeData = removeEdge === null || removeEdge === void 0 ? void 0 : removeEdge.getData();
                        return [4 /*yield*/, commandService.executeCommand(constant_1.XFlowEdgeCommands.DEL_EDGE.id, { edgeConfig: edgeData })];
                    case 17:
                        _g.sent();
                        _g.label = 18;
                    case 18:
                        _d++;
                        return [3 /*break*/, 16];
                    case 19:
                        _e = 0, updateNodes_1 = updateNodes;
                        _g.label = 20;
                    case 20:
                        if (!(_e < updateNodes_1.length)) return [3 /*break*/, 23];
                        updateNode = updateNodes_1[_e];
                        nodeData = updateNode === null || updateNode === void 0 ? void 0 : updateNode.getData();
                        return [4 /*yield*/, commandService.executeCommand(constant_1.XFlowNodeCommands.UPDATE_NODE.id, {
                                nodeConfig: nodeData,
                            })];
                    case 21:
                        _g.sent();
                        _g.label = 22;
                    case 22:
                        _e++;
                        return [3 /*break*/, 20];
                    case 23:
                        _f = 0, updateEdges_1 = updateEdges;
                        _g.label = 24;
                    case 24:
                        if (!(_f < updateEdges_1.length)) return [3 /*break*/, 27];
                        updateEdge = updateEdges_1[_f];
                        edgeData = updateEdge === null || updateEdge === void 0 ? void 0 : updateEdge.getData();
                        return [4 /*yield*/, commandService.executeCommand(constant_1.XFlowEdgeCommands.UPDATE_EDGE.id, {
                                edgeConfig: edgeData,
                            })];
                    case 25:
                        _g.sent();
                        _g.label = 26;
                    case 26:
                        _f++;
                        return [3 /*break*/, 24];
                    case 27:
                        if (x6Graph === null || x6Graph === void 0 ? void 0 : x6Graph.isFrozen()) {
                            x6Graph === null || x6Graph === void 0 ? void 0 : x6Graph.unfreeze();
                        }
                        return [2 /*return*/];
                }
            });
        }); };
    }
    GraphRenderCommand.prototype.init = function () {
        this.ctx = this.contextProvider();
    };
    GraphRenderCommand.prototype.isUndoable = function () {
        return this.ctx.isUndoable();
    };
    /**
     * 画布内容Diff
     * @param x6Graph x6画布实例
     * @param graphData 画布数据
     * @param isNodeEqual 允许用户自定义判断节点是否相等
     * @param isEdgeEqual 允许用户自定义判断边是否相等
     */
    GraphRenderCommand.prototype.graphDataDiff = function (x6Graph, graphData, isNodeEqual, isEdgeEqual) {
        var nodeConfigs = graphData.nodes, edgeConfigs = graphData.edges;
        /** 新增节点数据 */
        var addNodeConfigs = [];
        nodeConfigs.forEach(function (nodeConfig) {
            var findNode = x6Graph.getCellById(nodeConfig === null || nodeConfig === void 0 ? void 0 : nodeConfig.id);
            if (!findNode) {
                addNodeConfigs.push(nodeConfig);
            }
        });
        /** 保持、更新、移除节点 */
        var retainNodes = [];
        var updateNodes = [];
        var removeNodes = [];
        var allNodes = x6Graph.getNodes();
        allNodes.forEach(function (x6Node) {
            var findNodeConfig = nodeConfigs.find(function (nodeConfig) { return (nodeConfig === null || nodeConfig === void 0 ? void 0 : nodeConfig.id) === (x6Node === null || x6Node === void 0 ? void 0 : x6Node.id); });
            if (!findNodeConfig) {
                removeNodes.push(x6Node);
            }
            else {
                var judgeResult = true;
                if (isNodeEqual) {
                    /** 如果用户自定义节点是否相等的方法 */
                    judgeResult = isNodeEqual(x6Node === null || x6Node === void 0 ? void 0 : x6Node.data, findNodeConfig);
                }
                else {
                    /** XFlow默认的判断节点是否相等的逻辑 */
                    if ((x6Node === null || x6Node === void 0 ? void 0 : x6Node.data) && findNodeConfig) {
                        judgeResult = NsGraphUtils.isNodeEqual(x6Node === null || x6Node === void 0 ? void 0 : x6Node.data, findNodeConfig);
                    }
                }
                if (!judgeResult) {
                    x6Node.setData(findNodeConfig);
                }
                judgeResult === true ? retainNodes.push(x6Node) : updateNodes.push(x6Node);
            }
        });
        /** 新增边数据 */
        var addEdgeConfigs = [];
        edgeConfigs.forEach(function (edgeConfig) {
            var findEdge = x6Graph.getCellById(edgeConfig === null || edgeConfig === void 0 ? void 0 : edgeConfig.id);
            if (!findEdge) {
                addEdgeConfigs.push(edgeConfig);
            }
        });
        /** 保持、更新、移除节点 */
        var retainEdges = [];
        var updateEdges = [];
        var removeEdges = [];
        var allEdges = x6Graph.getEdges();
        allEdges.forEach(function (x6Edge) {
            var findEdgeConfig = edgeConfigs.find(function (edgeConfig) { return (edgeConfig === null || edgeConfig === void 0 ? void 0 : edgeConfig.id) === (x6Edge === null || x6Edge === void 0 ? void 0 : x6Edge.id); });
            if (!findEdgeConfig) {
                removeEdges.push(x6Edge);
            }
            else {
                var judgeResult = true;
                if (isEdgeEqual) {
                    /** 如果用户自定义边是否相等的方法 */
                    judgeResult = isEdgeEqual(x6Edge === null || x6Edge === void 0 ? void 0 : x6Edge.data, findEdgeConfig);
                }
                else {
                    /** XFlow默认的判断边是否相等的逻辑 */
                    if ((x6Edge === null || x6Edge === void 0 ? void 0 : x6Edge.data) && findEdgeConfig) {
                        judgeResult = NsGraphUtils.isEdgeEqual(x6Edge === null || x6Edge === void 0 ? void 0 : x6Edge.data, findEdgeConfig);
                    }
                }
                if (!judgeResult) {
                    x6Edge.setData(findEdgeConfig);
                }
                judgeResult === true ? retainEdges.push(x6Edge) : updateEdges.push(x6Edge);
            }
        });
        return {
            addNodeConfigs: addNodeConfigs,
            addEdgeConfigs: addEdgeConfigs,
            retainNodes: retainNodes,
            retainEdges: retainEdges,
            removeNodes: removeNodes,
            removeEdges: removeEdges,
            updateNodes: updateNodes,
            updateEdges: updateEdges,
        };
    };
    __decorate([
        (0, mana_syringe_1.inject)(interface_1.ICommandContextProvider),
        __metadata("design:type", Object)
    ], GraphRenderCommand.prototype, "contextProvider", void 0);
    __decorate([
        (0, mana_syringe_1.postConstruct)(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], GraphRenderCommand.prototype, "init", null);
    GraphRenderCommand = __decorate([
        (0, mana_syringe_1.injectable)({
            token: { token: interface_1.ICommandHandler, named: NsGraphRender.command.id },
        })
        /** 画布渲染命令 */
    ], GraphRenderCommand);
    return GraphRenderCommand;
}());
exports.GraphRenderCommand = GraphRenderCommand;
