// @antv/xflow-core/es/command-contributions/graph/graph-load-data.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { inject, injectable, postConstruct } from 'mana-syringe';
import { HookHub } from '@antv/xflow-hook';
import { ICommandHandler, ICommandContextProvider } from '../../command/interface';
import { XFlowGraphCommands } from '../constant';
/** 从服务端获取画布数据命令 */
export var NsGraphLoadData;
(function (NsGraphLoadData) {
    /** Command: 用于注册named factory */
    NsGraphLoadData.command = XFlowGraphCommands.LOAD_DATA;
    /** hookName */
    NsGraphLoadData.hookKey = 'loadData';
    NsGraphLoadData.createHook = () => {
        return new HookHub();
    };
})(NsGraphLoadData || (NsGraphLoadData = {}));
let GraphLoadDataCommand = 
/** 画布数据获取命令 */
class GraphLoadDataCommand {
    constructor() {
        /** 执行cmd */
        this.execute = () => __awaiter(this, void 0, void 0, function* () {
            const { args, hooks: runtimeHook } = this.ctx.getArgs();
            const hooks = this.ctx.getHooks();
            const result = yield hooks.loadData.call(args, (handlerArgs) => __awaiter(this, void 0, void 0, function* () {
                const { loadDataService, graphMeta } = handlerArgs;
                const graphData = yield loadDataService(graphMeta);
                return { graphData };
            }), runtimeHook);
            /** 设置结果 */
            this.ctx.setResult(result);
            return this;
        });
        this.undo = () => __awaiter(this, void 0, void 0, function* () {
            this.ctx.undo();
            return this;
        });
        this.redo = () => __awaiter(this, void 0, void 0, function* () {
            if (!this.ctx.isUndoable) {
                yield this.execute();
            }
            return this;
        });
    }
    init() {
        this.ctx = this.contextProvider();
    }
    isUndoable() {
        return this.ctx.isUndoable();
    }
};
__decorate([
    inject(ICommandContextProvider),
    __metadata("design:type", Object)
], GraphLoadDataCommand.prototype, "contextProvider", void 0);
__decorate([
    postConstruct(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], GraphLoadDataCommand.prototype, "init", null);
GraphLoadDataCommand = __decorate([
    injectable({
        token: { token: ICommandHandler, named: NsGraphLoadData.command.id },
    })
    /** 画布数据获取命令 */
], GraphLoadDataCommand);
export { GraphLoadDataCommand };
