// @antv/xflow-core/es/model-service/config.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { Deferred } from '../common/deferred';
import { Disposable } from '../common/disposable';
export var NsModelServiceConfig;
(function (NsModelServiceConfig) {
    NsModelServiceConfig.CONFIG_TYPE = 'MODEL_SERVICE_CONFIG';
})(NsModelServiceConfig || (NsModelServiceConfig = {}));
const dispoableNoop = () => Disposable.create(() => { });
export class ModelServiceConfig {
    constructor() {
        /** CONFIG_TYPE */
        this.CONFIG_TYPE = NsModelServiceConfig.CONFIG_TYPE;
        /** 外部注册context的方法 */
        this.registerModelFn = dispoableNoop;
        /** CONFIG_TYPE */
        this.isMounted = new Deferred();
        this.registerModel = (registerModel) => {
            this.registerModelFn = registerModel;
        };
        this.setMountState = () => {
            this.isMounted.resolve(true);
        };
        this.getConfig = () => __awaiter(this, void 0, void 0, function* () {
            yield this.isMounted.promise;
            return {
                CONFIG_TYPE: this.CONFIG_TYPE,
                modelRegisterFunc: this.registerModelFn || dispoableNoop,
            };
        });
        this.dispose = () => {
            this.registerModelFn = dispoableNoop;
            this.isMounted = new Deferred();
        };
    }
}
