// @antv/l7-core/lib/services/renderer/passes/BaseNormalPass.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _initializerDefineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/initializerDefineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _applyDecoratedDescriptor2 = _interopRequireDefault(require("@babel/runtime/helpers/applyDecoratedDescriptor"));

var _initializerWarningHelper2 = _interopRequireDefault(require("@babel/runtime/helpers/initializerWarningHelper"));

var _inversify = require("inversify");

require("reflect-metadata");

var _types = require("../../../types");

var _IMultiPassRenderer = require("../IMultiPassRenderer");

var _dec, _dec2, _class, _class2, _descriptor;

var BaseNormalPass = (_dec = (0, _inversify.injectable)(), _dec2 = (0, _inversify.inject)(_types.TYPES.IShaderModuleService), _dec(_class = (_class2 = function () {
  function BaseNormalPass() {
    (0, _classCallCheck2.default)(this, BaseNormalPass);
    (0, _initializerDefineProperty2.default)(this, "shaderModuleService", _descriptor, this);
    (0, _defineProperty2.default)(this, "rendererService", void 0);
    (0, _defineProperty2.default)(this, "cameraService", void 0);
    (0, _defineProperty2.default)(this, "mapService", void 0);
    (0, _defineProperty2.default)(this, "interactionService", void 0);
    (0, _defineProperty2.default)(this, "layerService", void 0);
    (0, _defineProperty2.default)(this, "config", void 0);
  }

  (0, _createClass2.default)(BaseNormalPass, [{
    key: "getName",
    value: function getName() {
      return '';
    }
  }, {
    key: "getType",
    value: function getType() {
      return _IMultiPassRenderer.PassType.Normal;
    }
  }, {
    key: "init",
    value: function init(layer, config) {
      this.config = config;
      this.rendererService = layer.getContainer().get(_types.TYPES.IRendererService);
      this.cameraService = layer.getContainer().get(_types.TYPES.ICameraService);
      this.mapService = layer.getContainer().get(_types.TYPES.IMapService);
      this.interactionService = layer.getContainer().get(_types.TYPES.IInteractionService);
      this.layerService = layer.getContainer().get(_types.TYPES.ILayerService);
    }
  }, {
    key: "render",
    value: function render(layer) {}
  }]);
  return BaseNormalPass;
}(), (_descriptor = (0, _applyDecoratedDescriptor2.default)(_class2.prototype, "shaderModuleService", [_dec2], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: null
})), _class2)) || _class);
exports.default = BaseNormalPass;
