// @antv/xflow-core/es/command-contributions/edge/edge-highlight.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { inject, injectable, postConstruct } from 'mana-syringe';
import { ICommandHandler, ICommandContextProvider } from '../../command/interface';
import { XFlowEdgeCommands } from '../constant';
export var NsHighlightEdge;
(function (NsHighlightEdge) {
    NsHighlightEdge.command = XFlowEdgeCommands.HIGHLIGHT_EDGE;
    NsHighlightEdge.hookKey = 'highlightEdge';
})(NsHighlightEdge || (NsHighlightEdge = {}));
let HighlightEdgeCommand = 
/** 连线高亮 */
class HighlightEdgeCommand {
    constructor() {
        this.execute = () => __awaiter(this, void 0, void 0, function* () {
            const { args, hooks: runtimeHook } = this.ctx.getArgs();
            const hooks = this.ctx.getHooks();
            const result = yield hooks.highlightEdge.call(args, (handlerArgs) => __awaiter(this, void 0, void 0, function* () {
                var _a, _b;
                const x6Graph = yield this.ctx.getX6Graph();
                const { edgeId, strokeColor, strokeWidth } = handlerArgs;
                const allEdges = x6Graph.getEdges();
                /** 需要高亮的连线 */
                const highlightEdge = allEdges.find(edge => edge.id === edgeId);
                /** 不需要高亮的连线 */
                const otherEdges = allEdges.filter(edge => edge.id !== edgeId);
                if (!highlightEdge) {
                    console.error(edgeId, 'this edgeId is not exist');
                }
                else {
                    const oldAttr = highlightEdge.getAttrs();
                    if (((_a = oldAttr === null || oldAttr === void 0 ? void 0 : oldAttr.line) === null || _a === void 0 ? void 0 : _a.stroke) === strokeColor && ((_b = oldAttr === null || oldAttr === void 0 ? void 0 : oldAttr.line) === null || _b === void 0 ? void 0 : _b.strokeWidth) === strokeWidth) {
                        /** 连线已经高亮, 不需要重复操作 */
                    }
                    else {
                        /** 高亮选中的连线 */
                        highlightEdge === null || highlightEdge === void 0 ? void 0 : highlightEdge.setAttrs({
                            line: {
                                stroke: strokeColor,
                                strokeWidth: strokeWidth || 2,
                            },
                        });
                        /** 其余连线取消高亮状态 */
                        otherEdges.forEach(edge => {
                            edge.setAttrs({
                                line: oldAttr === null || oldAttr === void 0 ? void 0 : oldAttr.line,
                            });
                        });
                        /** 高亮的连线默认前置在画布最前方 */
                        handlerArgs.commandService.executeCommand(XFlowEdgeCommands.FRONT_EDGE.id, {
                            edgeId,
                        });
                    }
                }
                return {};
            }), runtimeHook);
            this.ctx.setResult(result);
            return this;
        });
        this.undo = () => __awaiter(this, void 0, void 0, function* () {
            this.ctx.undo();
            return this;
        });
        this.redo = () => __awaiter(this, void 0, void 0, function* () {
            if (!this.ctx.isUndoable) {
                yield this.execute();
            }
            return this;
        });
    }
    init() {
        this.ctx = this.contextProvider();
    }
    isUndoable() {
        return this.ctx.isUndoable();
    }
};
__decorate([
    inject(ICommandContextProvider),
    __metadata("design:type", Object)
], HighlightEdgeCommand.prototype, "contextProvider", void 0);
__decorate([
    postConstruct(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], HighlightEdgeCommand.prototype, "init", null);
HighlightEdgeCommand = __decorate([
    injectable({
        token: { token: ICommandHandler, named: NsHighlightEdge.command.id },
    })
    /** 连线高亮 */
], HighlightEdgeCommand);
export { HighlightEdgeCommand };
