// @antv/xflow-extension/lib/canvas-context-menu/components/menu-render.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderMenuOptions = void 0;
var react_1 = __importDefault(require("react"));
var xflow_core_1 = require("@antv/xflow-core");
var x6_react_components_1 = require("@antv/x6-react-components");
var renderMenuOptions = function (props) {
    var idx = props.idx, menuItem = props.menuOptions, target = props.target, modelService = props.modelService, commandService = props.commandService, onHide = props.onHide;
    var id = menuItem.id, label = menuItem.label, _a = menuItem.isEnabled, isEnabled = _a === void 0 ? true : _a, _b = menuItem.isVisible, isVisible = _b === void 0 ? true : _b, iconName = menuItem.iconName, _c = menuItem.submenu, submenu = _c === void 0 ? [] : _c, active = menuItem.active, hotkey = menuItem.hotkey, render = menuItem.render, onClick = menuItem.onClick;
    if (!isVisible) {
        return null;
    }
    /** 分隔符：separator */
    if (menuItem.type === 'separator') {
        return react_1.default.createElement(x6_react_components_1.Menu.Divider, { key: String(menuItem.id) + idx });
    }
    /** 根结点：root */
    if (menuItem.type === 'root') {
        return (react_1.default.createElement(x6_react_components_1.Menu, { hasIcon: true }, submenu.map(function (item, submenuIdx) {
            return (0, exports.renderMenuOptions)({
                idx: submenuIdx,
                target: target,
                commandService: commandService,
                modelService: modelService,
                menuOptions: item,
                onHide: onHide,
            });
        })));
    }
    /** 子菜单：submenu */
    if (menuItem.submenu && Array.isArray(menuItem.submenu)) {
        var Icon_1 = xflow_core_1.IconStore.get(iconName);
        var isDisable = !isEnabled;
        return (react_1.default.createElement(x6_react_components_1.Menu.SubMenu, { key: String(id), text: label, disabled: isDisable || submenu.length === 0, icon: Icon_1 ? react_1.default.createElement(Icon_1, null) : null, active: active, hotkey: hotkey }, submenu.map(function (item, submenuIdx) {
            return (0, exports.renderMenuOptions)({
                idx: submenuIdx,
                target: target,
                commandService: commandService,
                modelService: modelService,
                menuOptions: item,
                onHide: onHide,
            });
        })));
    }
    /** 叶子节点 */
    var Icon = xflow_core_1.IconStore.get(iconName);
    if (render) {
        return react_1.default.createElement(render, { menuItem: menuItem, target: target, commandService: commandService, modelService: modelService, onHide: onHide }, [
            react_1.default.createElement(x6_react_components_1.Menu.Item, { key: String(id), text: label, disabled: !isEnabled, icon: Icon ? react_1.default.createElement(Icon, null) : null, active: active, hotkey: hotkey }),
        ]);
    }
    return (react_1.default.createElement(x6_react_components_1.Menu.Item, { key: String(id), text: label, disabled: !isEnabled, icon: Icon ? react_1.default.createElement(Icon, null) : null, active: active, hotkey: hotkey, onClick: function () { return __awaiter(void 0, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, onClick({ menuItem: menuItem, target: target, commandService: commandService, modelService: modelService })];
                    case 1:
                        _a.sent();
                        onHide && onHide();
                        return [2 /*return*/];
                }
            });
        }); } }));
};
exports.renderMenuOptions = renderMenuOptions;
