// @antv/xflow-extension/es/canvas-toolbar/service.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { useXFlowApp, createComponentModel, DisposableCollection } from '@antv/xflow-core';
import React from 'react';
export const useToolbarModel = (props) => {
    const { config } = props;
    const modelId = config.moduleId;
    const app = useXFlowApp();
    const [state, setState, toolbarModel, isModelReady] = createComponentModel({
        name: modelId,
        layout: props.layout,
        mainGroups: [],
        extraGroups: [],
        customRender: null,
    });
    /** 注册全局的model */
    React.useEffect(() => {
        const toDispose = new DisposableCollection();
        const deferredModel = app.modelService.findDeferredModel(modelId);
        if (!deferredModel) {
            const { toolbarCustomRender, toolbarModelService } = config.getConfig();
            const d = app.modelService.registerModel({
                id: modelId,
                modelFactory: () => toolbarModel,
                watchChange: (self) => __awaiter(void 0, void 0, void 0, function* () {
                    if (toolbarModelService) {
                        yield toolbarModelService(self, app.modelService, toDispose);
                    }
                    if (toolbarCustomRender) {
                        const updateCustomRender = customRender => {
                            self.setValue(m => (m.customRender = customRender));
                        };
                        yield toolbarCustomRender(app.modelService, updateCustomRender);
                    }
                    return toDispose;
                }),
            });
            toDispose.push(d);
        }
        return () => {
            toDispose.dispose();
        };
        /* eslint-disable-next-line  */
    }, []);
    return { isModelReady, state, setState, toolbarModel };
};
